// egov-common-utility/table/ServerSideTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _Card = require('material-ui/Card');

var _reactBootstrap = require('react-bootstrap');

var _reactPaginate = require('react-paginate');

var _reactPaginate2 = _interopRequireDefault(_reactPaginate);

var _materialUi = require('../styles/material-ui');

var _materialUi2 = _interopRequireDefault(_materialUi);

var _common = require('../common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ServerSideTable = function (_Component) {
  _inherits(ServerSideTable, _Component);

  function ServerSideTable() {
    _classCallCheck(this, ServerSideTable);

    var _this = _possibleConstructorReturn(this, (ServerSideTable.__proto__ || Object.getPrototypeOf(ServerSideTable)).call(this));

    _this.handleNavigation = function (path) {
      _this.props.setRoute(path);
    };

    _this.handleChange = function (event) {
      var results = [];
      var toSearch = event.target.value.toLowerCase();
      for (var i = 0; i < _this.state.resultList.reportReponseData.length; i++) {
        for (var key in _this.state.resultList.reportReponseData[i]) {
          if (_this.state.resultList.reportReponseData[i][key].toLowerCase().indexOf(toSearch) != -1) {
            results.push(_this.state.resultList.reportReponseData[i]);
          }
        }
      }
      _this.setState({
        userSearch: true,
        value: event.target.value,
        userSearchResult: results
      });
    };

    _this.renderHeader = function () {
      var resultList = _this.state.resultList;

      if (resultList) {
        if (resultList.reportReponseData && resultList.reportReponseData.length > 0) {
          return Object.keys(resultList.reportReponseData[0]).map(function (header, index) {
            return _react2.default.createElement(
              'th',
              { key: index },
              header
            );
          });
        }
      }
    };

    _this.renderBody = function () {
      var _this$state = _this.state,
          resultList = _this$state.resultList,
          userSearchResult = _this$state.userSearchResult,
          userSearch = _this$state.userSearch;

      if (resultList) {
        var searchResponse = userSearch ? userSearchResult : resultList.reportReponseData;
        if (searchResponse && searchResponse.length > 0) {
          return searchResponse.map(function (obj, index) {
            return _react2.default.createElement(
              'tr',
              { key: index },
              Object.values(obj).map(function (data, idx) {
                if (Object.values(resultList.reportActionData[0])[idx].trim()) {
                  return _react2.default.createElement(
                    'td',
                    { key: idx },
                    _react2.default.createElement(
                      'a',
                      { href: 'javascript:void(0)', onClick: function onClick(e) {
                          _this.handleNavigation(Object.values(resultList.reportActionData[0])[idx] + encodeURIComponent(data));
                        } },
                      data
                    )
                  );
                } else {
                  return _react2.default.createElement(
                    'td',
                    { key: idx },
                    data
                  );
                }
              })
            );
          });
        }
      }
    };

    _this.showPagination = function () {
      return _react2.default.createElement(
        _reactBootstrap.Col,
        { md: 12 },
        _react2.default.createElement(
          'div',
          { style: { "textAlign": "center" } },
          _react2.default.createElement(_reactPaginate2.default, { previousLabel: (0, _common.translate)("pgr.lbl.previous"),
            nextLabel: (0, _common.translate)("pgr.lbl.next"),
            breakLabel: _react2.default.createElement(
              'a',
              { href: '' },
              '...'
            ),
            pageCount: _this.state.pageCount,
            marginPagesDisplayed: 2,
            pageRangeDisplayed: 5,
            onPageChange: _this.handlePageClick,
            containerClassName: "pagination",
            subContainerClassName: "pages pagination",
            activeClassName: "active" })
        )
      );
    };

    _this.handlePageClick = function (data) {
      var selected = data.selected;
      var offset = Math.ceil(selected * 10);
      _this.props.search(true, offset);
    };

    _this.state = {};
    return _this;
  }

  _createClass(ServerSideTable, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.setState({
        resultList: this.props.resultSet,
        pageCount: this.props.pageCount,
        userSearch: false
      });
      value: '';
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.props.resultSet !== nextProps.resultSet) {
        this.setState({
          resultList: nextProps.resultSet,
          pageCount: nextProps.pageCount,
          userSearch: false,
          value: ''
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _Card.Card,
        { style: _materialUi2.default.marginStyle },
        _react2.default.createElement(_Card.CardHeader, { style: { paddingBottom: 0 }, title: _react2.default.createElement(
            'div',
            { style: _materialUi2.default.headerStyle },
            ' ',
            (0, _common.translate)("pgr.searchresult"),
            ' '
          ) }),
        _react2.default.createElement(
          _Card.CardText,
          { style: { paddingTop: 0 } },
          _react2.default.createElement(
            _reactBootstrap.Row,
            null,
            _react2.default.createElement(
              _reactBootstrap.Col,
              { mdOffset: 7, md: 3 },
              _react2.default.createElement(
                'label',
                { className: 'pull-right' },
                'Search: '
              )
            ),
            _react2.default.createElement(
              _reactBootstrap.Col,
              { md: 2 },
              _react2.default.createElement('input', { type: 'text', className: 'form-control', value: this.state.userSearch ? this.state.value : '', onChange: this.handleChange, style: { marginBottom: 5 } })
            ),
            _react2.default.createElement(
              _reactBootstrap.Col,
              { md: 12 },
              _react2.default.createElement(
                _reactBootstrap.Table,
                { style: { color: "black", fontWeight: "normal" }, bordered: true, responsive: true, className: 'table-striped' },
                _react2.default.createElement(
                  'thead',
                  null,
                  _react2.default.createElement(
                    'tr',
                    null,
                    this.renderHeader()
                  )
                ),
                _react2.default.createElement(
                  'tbody',
                  null,
                  this.renderBody()
                )
              )
            ),
            this.showPagination()
          )
        )
      );
    }
  }]);

  return ServerSideTable;
}(_react.Component);

// export default ServerSideTable;


var mapStateToProps = function mapStateToProps(state) {
  return {};
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
  return {
    setRoute: function setRoute(route) {
      return dispatch({ type: "SET_ROUTE", route: route });
    }
  };
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(ServerSideTable);
