// egov-common-utility/ImagePreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrap = require('react-bootstrap');

var _FlatButton = require('material-ui/FlatButton');

var _FlatButton2 = _interopRequireDefault(_FlatButton);

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _highlightOff = require('material-ui/svg-icons/action/highlight-off');

var _highlightOff2 = _interopRequireDefault(_highlightOff);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var filename;
var renderImage = function renderImage(props) {

  var readURL = function readURL(e, idx) {
    if (props.file) {
      var section = 'image' + idx;
      var reader = new FileReader();
      reader.onload = function (e) {
        document.getElementById(section).src = e.target.result;
      };
      reader.readAsDataURL(props.file);
    }
  };

  var removeFile = function removeFile(fileobjToDelete) {
    props.handler(fileobjToDelete.name);
  };

  return _react2.default.createElement(
    _reactBootstrap.Col,
    { xs: 12, md: 3 },
    _react2.default.createElement(_RaisedButton2.default, {
      label: props.file.name.length > 15 ? props.file.name.substr(0, 12) + '...' : props.file.name,
      primary: true,
      fullWidth: true,
      icon: _react2.default.createElement(_highlightOff2.default, null),
      onTouchTap: function onTouchTap(event) {
        removeFile(props.file);
      }
    }),
    _react2.default.createElement('img', { src: '#', id: 'image' + props.idx, width: '100%', height: '100px', style: { objectFit: 'cover', cursor: 'pointer', marginBottom: '20px' }, onClick: function onClick(e) {
        props.preview(document.getElementById('image' + props.idx).src);
      } }),
    readURL(undefined, props.idx)
  );
};

exports.default = renderImage;
