// egov-ui-react-framework/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.UiBackButton = exports.UiWindowForm = exports.UiDynamicTable2 = exports.UiDynamicTable = exports.getTitleCase = exports.getInitiatorPosition = exports.fileUpload = exports.getFullDate = exports.int_to_words = exports.UiCalendar = exports.UiTimeField = exports.UiWorkflow = exports.UigoogleMaps = exports.UiMultiFieldTable = exports.UiFileTable = exports.UiArrayField = exports.UiPinCode = exports.UiAutoComplete = exports.UiDocumentList = exports.UiTextSearch = exports.UiRadioButton = exports.UiLabel = exports.UiPanCard = exports.UiAadharCard = exports.UiSingleFileUpload = exports.UiMultiFileUpload = exports.UiDatePicker = exports.UiNumberField = exports.UiMultiSelectField = exports.UiTextArea = exports.UiMobileNumber = exports.UiEmailField = exports.UiCheckBox = exports.UiButton = exports.UiSelectFieldMultiple = exports.UiSelectField = exports.ShowFields = exports.UiTextField = exports.routes = undefined;

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

var _view = require('./view');

var _view2 = _interopRequireDefault(_view);

var _search = require('./search');

var _search2 = _interopRequireDefault(_search);

var _showFields = require('./showFields');

var _showFields2 = _interopRequireDefault(_showFields);

var _UiTextField = require('./components/UiTextField');

var _UiTextField2 = _interopRequireDefault(_UiTextField);

var _UiSelectField = require('./components/UiSelectField');

var _UiSelectField2 = _interopRequireDefault(_UiSelectField);

var _UiSelectFieldMultiple = require('./components/UiSelectFieldMultiple');

var _UiSelectFieldMultiple2 = _interopRequireDefault(_UiSelectFieldMultiple);

var _UiButton = require('./components/UiButton');

var _UiButton2 = _interopRequireDefault(_UiButton);

var _UiCheckBox = require('./components/UiCheckBox');

var _UiCheckBox2 = _interopRequireDefault(_UiCheckBox);

var _UiEmailField = require('./components/UiEmailField');

var _UiEmailField2 = _interopRequireDefault(_UiEmailField);

var _UiMobileNumber = require('./components/UiMobileNumber');

var _UiMobileNumber2 = _interopRequireDefault(_UiMobileNumber);

var _UiTextArea = require('./components/UiTextArea');

var _UiTextArea2 = _interopRequireDefault(_UiTextArea);

var _UiMultiSelectField = require('./components/UiMultiSelectField');

var _UiMultiSelectField2 = _interopRequireDefault(_UiMultiSelectField);

var _UiNumberField = require('./components/UiNumberField');

var _UiNumberField2 = _interopRequireDefault(_UiNumberField);

var _UiDatePicker = require('./components/UiDatePicker');

var _UiDatePicker2 = _interopRequireDefault(_UiDatePicker);

var _UiMultiFileUpload = require('./components/UiMultiFileUpload');

var _UiMultiFileUpload2 = _interopRequireDefault(_UiMultiFileUpload);

var _UiSingleFileUpload = require('./components/UiSingleFileUpload');

var _UiSingleFileUpload2 = _interopRequireDefault(_UiSingleFileUpload);

var _UiAadharCard = require('./components/UiAadharCard');

var _UiAadharCard2 = _interopRequireDefault(_UiAadharCard);

var _UiPanCard = require('./components/UiPanCard');

var _UiPanCard2 = _interopRequireDefault(_UiPanCard);

var _UiLabel = require('./components/UiLabel');

var _UiLabel2 = _interopRequireDefault(_UiLabel);

var _UiRadioButton = require('./components/UiRadioButton');

var _UiRadioButton2 = _interopRequireDefault(_UiRadioButton);

var _UiTextSearch = require('./components/UiTextSearch');

var _UiTextSearch2 = _interopRequireDefault(_UiTextSearch);

var _UiDocumentList = require('./components/UiDocumentList');

var _UiDocumentList2 = _interopRequireDefault(_UiDocumentList);

var _UiAutoComplete = require('./components/UiAutoComplete');

var _UiAutoComplete2 = _interopRequireDefault(_UiAutoComplete);

var _UiPinCode = require('./components/UiPinCode');

var _UiPinCode2 = _interopRequireDefault(_UiPinCode);

var _UiArrayField = require('./components/UiArrayField');

var _UiArrayField2 = _interopRequireDefault(_UiArrayField);

var _UiFileTable = require('./components/UiFileTable');

var _UiFileTable2 = _interopRequireDefault(_UiFileTable);

var _UiMultiFieldTable = require('./components/UiMultiFieldTable');

var _UiMultiFieldTable2 = _interopRequireDefault(_UiMultiFieldTable);

var _UigoogleMaps = require('./components/UigoogleMaps');

var _UigoogleMaps2 = _interopRequireDefault(_UigoogleMaps);

var _UiWorkflow = require('./components/UiWorkflow');

var _UiWorkflow2 = _interopRequireDefault(_UiWorkflow);

var _UiTimeField = require('./components/UiTimeField');

var _UiTimeField2 = _interopRequireDefault(_UiTimeField);

var _UiCalendar = require('./components/UiCalendar');

var _UiCalendar2 = _interopRequireDefault(_UiCalendar);

var _UiDynamicTable = require('./components/UiDynamicTable');

var _UiDynamicTable2 = _interopRequireDefault(_UiDynamicTable);

var _uiDynamicTable = require('./components/uiDynamicTable2');

var _uiDynamicTable2 = _interopRequireDefault(_uiDynamicTable);

var _UiWindowForm = require('./components/UiWindowForm');

var _UiWindowForm2 = _interopRequireDefault(_UiWindowForm);

var _UiBackButton = require('./components/UiBackButton');

var _UiBackButton2 = _interopRequireDefault(_UiBackButton);

var _templateParser = require('./templates/templateParser/templateParser');

var _templateParser2 = _interopRequireDefault(_templateParser);

var _utility = require('./utility/utility');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var routes = [{
	component: _create2.default,
	route: '/create/:moduleName/:master?/:id?'
}, {
	component: _view2.default,
	route: '/view/:moduleName/:master?/:id'
}, {
	component: _search2.default,
	route: '/search/:moduleName/:master?/:action'
}, {
	component: _create2.default,
	route: '/update/:moduleName/:master?/:id?'
}, {
	component: _templateParser2.default,
	route: '/print/report/:templatePath'
}];

exports.routes = routes;
exports.UiTextField = _UiTextField2.default;
exports.ShowFields = _showFields2.default;
exports.UiSelectField = _UiSelectField2.default;
exports.UiSelectFieldMultiple = _UiSelectFieldMultiple2.default;
exports.UiButton = _UiButton2.default;
exports.UiCheckBox = _UiCheckBox2.default;
exports.UiEmailField = _UiEmailField2.default;
exports.UiMobileNumber = _UiMobileNumber2.default;
exports.UiTextArea = _UiTextArea2.default;
exports.UiMultiSelectField = _UiMultiSelectField2.default;
exports.UiNumberField = _UiNumberField2.default;
exports.UiDatePicker = _UiDatePicker2.default;
exports.UiMultiFileUpload = _UiMultiFileUpload2.default;
exports.UiSingleFileUpload = _UiSingleFileUpload2.default;
exports.UiAadharCard = _UiAadharCard2.default;
exports.UiPanCard = _UiPanCard2.default;
exports.UiLabel = _UiLabel2.default;
exports.UiRadioButton = _UiRadioButton2.default;
exports.UiTextSearch = _UiTextSearch2.default;
exports.UiDocumentList = _UiDocumentList2.default;
exports.UiAutoComplete = _UiAutoComplete2.default;
exports.UiPinCode = _UiPinCode2.default;
exports.UiArrayField = _UiArrayField2.default;
exports.UiFileTable = _UiFileTable2.default;
exports.UiMultiFieldTable = _UiMultiFieldTable2.default;
exports.UigoogleMaps = _UigoogleMaps2.default;
exports.UiWorkflow = _UiWorkflow2.default;
exports.UiTimeField = _UiTimeField2.default;
exports.UiCalendar = _UiCalendar2.default;
exports.int_to_words = _utility.int_to_words;
exports.getFullDate = _utility.getFullDate;
exports.fileUpload = _utility.fileUpload;
exports.getInitiatorPosition = _utility.getInitiatorPosition;
exports.getTitleCase = _utility.getTitleCase;
exports.UiDynamicTable = _UiDynamicTable2.default;
exports.UiDynamicTable2 = _uiDynamicTable2.default;
exports.UiWindowForm = _UiWindowForm2.default;
exports.UiBackButton = _UiBackButton2.default;
