// egov-ui-react-framework/showFields.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrap = require('react-bootstrap');

var _Card = require('material-ui/Card');

var _UiTextField = require('./components/UiTextField');

var _UiTextField2 = _interopRequireDefault(_UiTextField);

var _UiSelectField = require('./components/UiSelectField');

var _UiSelectField2 = _interopRequireDefault(_UiSelectField);

var _UiSelectFieldMultiple = require('./components/UiSelectFieldMultiple');

var _UiSelectFieldMultiple2 = _interopRequireDefault(_UiSelectFieldMultiple);

var _UiButton = require('./components/UiButton');

var _UiButton2 = _interopRequireDefault(_UiButton);

var _UiCheckBox = require('./components/UiCheckBox');

var _UiCheckBox2 = _interopRequireDefault(_UiCheckBox);

var _UiEmailField = require('./components/UiEmailField');

var _UiEmailField2 = _interopRequireDefault(_UiEmailField);

var _UiMobileNumber = require('./components/UiMobileNumber');

var _UiMobileNumber2 = _interopRequireDefault(_UiMobileNumber);

var _UiTextArea = require('./components/UiTextArea');

var _UiTextArea2 = _interopRequireDefault(_UiTextArea);

var _UiMultiSelectField = require('./components/UiMultiSelectField');

var _UiMultiSelectField2 = _interopRequireDefault(_UiMultiSelectField);

var _UiNumberField = require('./components/UiNumberField');

var _UiNumberField2 = _interopRequireDefault(_UiNumberField);

var _UiDatePicker = require('./components/UiDatePicker');

var _UiDatePicker2 = _interopRequireDefault(_UiDatePicker);

var _UiMultiFileUpload = require('./components/UiMultiFileUpload');

var _UiMultiFileUpload2 = _interopRequireDefault(_UiMultiFileUpload);

var _UiSingleFileUpload = require('./components/UiSingleFileUpload');

var _UiSingleFileUpload2 = _interopRequireDefault(_UiSingleFileUpload);

var _UiAadharCard = require('./components/UiAadharCard');

var _UiAadharCard2 = _interopRequireDefault(_UiAadharCard);

var _UiPanCard = require('./components/UiPanCard');

var _UiPanCard2 = _interopRequireDefault(_UiPanCard);

var _UiLabel = require('./components/UiLabel');

var _UiLabel2 = _interopRequireDefault(_UiLabel);

var _UiRadioButton = require('./components/UiRadioButton');

var _UiRadioButton2 = _interopRequireDefault(_UiRadioButton);

var _UiTextSearch = require('./components/UiTextSearch');

var _UiTextSearch2 = _interopRequireDefault(_UiTextSearch);

var _UiDocumentList = require('./components/UiDocumentList');

var _UiDocumentList2 = _interopRequireDefault(_UiDocumentList);

var _UiAutoComplete = require('./components/UiAutoComplete');

var _UiAutoComplete2 = _interopRequireDefault(_UiAutoComplete);

var _FloatingActionButton = require('material-ui/FloatingActionButton');

var _FloatingActionButton2 = _interopRequireDefault(_FloatingActionButton);

var _UiDate = require('./components/UiDate');

var _UiDate2 = _interopRequireDefault(_UiDate);

var _UiPinCode = require('./components/UiPinCode');

var _UiPinCode2 = _interopRequireDefault(_UiPinCode);

var _UiArrayField = require('./components/UiArrayField');

var _UiArrayField2 = _interopRequireDefault(_UiArrayField);

var _UiFileTable = require('./components/UiFileTable');

var _UiFileTable2 = _interopRequireDefault(_UiFileTable);

var _UiMultiFieldTable = require('./components/UiMultiFieldTable');

var _UiMultiFieldTable2 = _interopRequireDefault(_UiMultiFieldTable);

var _UiHyperLink = require('./components/UiHyperLink');

var _UiHyperLink2 = _interopRequireDefault(_UiHyperLink);

var _UigoogleMaps = require('./components/UigoogleMaps');

var _UigoogleMaps2 = _interopRequireDefault(_UigoogleMaps);

var _UiWorkflow = require('./components/UiWorkflow');

var _UiWorkflow2 = _interopRequireDefault(_UiWorkflow);

var _UiTimeField = require('./components/UiTimeField');

var _UiTimeField2 = _interopRequireDefault(_UiTimeField);

var _UiWindowForm = require('./components/UiWindowForm');

var _UiWindowForm2 = _interopRequireDefault(_UiWindowForm);

var _UiCalendar = require('./components/UiCalendar');

var _UiCalendar2 = _interopRequireDefault(_UiCalendar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  reducePadding: {
    paddingTop: 4,
    paddingBottom: 0
  }
};

var ShowFields = function (_Component) {
  _inherits(ShowFields, _Component);

  function ShowFields(props) {
    _classCallCheck(this, ShowFields);

    var _this = _possibleConstructorReturn(this, (ShowFields.__proto__ || Object.getPrototypeOf(ShowFields)).call(this, props));

    _this.changeExpanded = function (name) {
      _this.setState(_defineProperty({}, name, !_this.state[name]));
    };

    _this.renderCard = function (group, groupIndex, noCols, jsonPath, uiFramework, groups, isMultiple) {
      var self = _this;
      var _this$props = _this.props,
          addNewCard = _this$props.addNewCard,
          removeCard = _this$props.removeCard;
      var renderField = _this.renderField;

      return _react2.default.createElement(
        _Card.Card,
        { style: { "display": eval(group.hide) ? "none" : "block", "marginBottom": isMultiple ? '0px' : '', "marginTop": isMultiple ? '0px' : '' }, className: "uiCard " + group.name, key: groupIndex, expanded: self.state[group.name] ? false : true, onExpandChange: function onExpandChange() {
            self.changeExpanded(group.name);
          } },
        !isMultiple && _react2.default.createElement(_Card.CardHeader, { style: { paddingTop: 4, paddingBottom: 0 }, title: _react2.default.createElement(
            'div',
            { style: { color: "#354f57", fontSize: 18, margin: '8px 0' } },
            group.label
          ), actAsExpander: true }),
        _react2.default.createElement(
          _Card.CardText,
          { style: { paddingTop: 0, paddingBottom: 0 }, expandable: true },
          _react2.default.createElement(
            _reactBootstrap.Grid,
            { style: { paddingTop: 0 } },
            _react2.default.createElement(
              _reactBootstrap.Row,
              null,
              group.fields.map(function (field, fieldIndex) {
                if (!field.isHidden) {
                  return _react2.default.createElement(
                    _reactBootstrap.Col,
                    { key: fieldIndex, xs: 12, sm: field.type === "documentList" || field.type === "fileTable" || field.type === "tableList" || field.type === "textarea" && field.fullWidth === true || field.type === "workflow" ? 12 : noCols, md: field.type === "documentList" || field.type === "fileTable" || field.type === "tableList" || field.type === "textarea" && field.fullWidth === true || field.type === "window" || field.type === "workflow" ? 12 : noCols },
                    renderField(field, self.props.screen, fieldIndex, field.screenView)
                  );
                }
              })
            ),
            self.props.screen != "view" && group.multiple && (!groups[groupIndex + 1] || groups[groupIndex + 1] && groups[groupIndex + 1].name != group.name) && _react2.default.createElement(
              _reactBootstrap.Row,
              null,
              _react2.default.createElement(
                _reactBootstrap.Col,
                { xsOffset: 8, mdOffset: 10, xs: 4, md: 2, style: { "textAlign": "right" } },
                _react2.default.createElement(
                  _FloatingActionButton2.default,
                  { mini: true, onClick: function onClick() {
                      addNewCard(group, jsonPath, group.name);
                    } },
                  _react2.default.createElement('span', { className: 'glyphicon glyphicon-plus' })
                )
              )
            ),
            self.props.screen != "view" && group.multiple && groups[groupIndex + 1] && groups[groupIndex + 1].name == group.name && _react2.default.createElement(
              _reactBootstrap.Row,
              null,
              _react2.default.createElement(
                _reactBootstrap.Col,
                { xsOffset: 8, mdOffset: 10, xs: 4, md: 2, style: { "textAlign": "right" } },
                _react2.default.createElement(
                  _FloatingActionButton2.default,
                  { mini: true, secondary: true, onClick: function onClick() {
                      removeCard(jsonPath, groupIndex, group.name);
                    } },
                  _react2.default.createElement('span', { className: 'glyphicon glyphicon-minus' })
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { style: { "marginLeft": "15px", "marginRight": "15px" } },
            group.children && group.children.length ? group.children.map(function (child) {
              return self.renderGroups(child.groups, noCols, uiFramework, child.jsonPath);
            }) : ""
          )
        )
      );
    };

    _this.renderGroups = function (groups, noCols) {
      var uiFramework = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "google";
      var jsonPath = arguments[3];

      var self = _this;
      switch (uiFramework) {
        case "google":
          var listArr = {};
          for (var i = 0; i < groups.length; i++) {
            if (groups[i].multiple) {
              if (!listArr[groups[i].name]) listArr[groups[i].name] = {
                objects: []
              };
              listArr[groups[i].name].objects.push({
                object: groups[i],
                index: i
              });
            } else {
              listArr[groups[i].name] = {
                object: groups[i],
                index: i
              };
            }
          }

          return Object.keys(listArr).map(function (key, groupIndex) {
            if (listArr[key].objects) {
              var _React$createElement;

              return _react2.default.createElement(
                _Card.Card,
                { className: "uiCard ", expanded: true },
                _react2.default.createElement(_Card.CardHeader, { style: { paddingTop: 4, paddingBottom: 0 }, title: _react2.default.createElement(
                    'div',
                    { style: { color: "#354f57", fontSize: 18, margin: '8px 0' } },
                    listArr[key].objects[0].object.label
                  ), subtitle: _typeof(listArr[key].objects[0].object.description) ? listArr[key].objects[0].object.description : "", showExpandableButton: true, actAsExpander: true }),
                _react2.default.createElement(
                  _Card.CardText,
                  (_React$createElement = { style: { paddingTop: 0, paddingBottom: 0 } }, _defineProperty(_React$createElement, 'style', { padding: 0 }), _defineProperty(_React$createElement, 'expandable', true), _React$createElement),
                  listArr[key].objects.map(function (grp, grpIndex) {
                    return self.renderCard(grp.object, grp.index, noCols, jsonPath, uiFramework, groups, true);
                  })
                )
              );
            } else {
              return self.renderCard(listArr[key].object, listArr[key].index, noCols, jsonPath, uiFramework, groups);
            }
          });
          break;
      }
    };

    _this.renderField = function (item, screen, index, screenView) {
      if (screen == "view" && ["documentList", "fileTable", "arrayText", "arrayNumber", "tableList", "workflow"].indexOf(item.type) == -1) {
        if (item.type == "datePicker") {
          item.isDate = true;
        }
        item.type = "label";
      }
      switch (item.type) {
        case 'text':
          return _react2.default.createElement(_UiTextField2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'singleValueListMultiple':
          return _react2.default.createElement(_UiSelectFieldMultiple2.default, { tabIndex: index, ui: _this.props.ui, useTimestamp: _this.props.useTimestamp, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'singleValueList':
          return _react2.default.createElement(_UiSelectField2.default, { tabIndex: index, ui: _this.props.ui, useTimestamp: _this.props.useTimestamp, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'multiValueList':
          return _react2.default.createElement(_UiMultiSelectField2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'autoCompelete':
          return _react2.default.createElement(_UiAutoComplete2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, autoComHandler: _this.props.autoComHandler || "" });
        case 'number':
          return _react2.default.createElement(_UiNumberField2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'textarea':
          return _react2.default.createElement(_UiTextArea2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'mobileNumber':
          return _react2.default.createElement(_UiMobileNumber2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'checkbox':
          return _react2.default.createElement(_UiCheckBox2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'email':
          return _react2.default.createElement(_UiEmailField2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'button':
          return _react2.default.createElement(_UiButton2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'datePicker':
          return _react2.default.createElement(_UiDatePicker2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'date':
          return _react2.default.createElement(_UiDate2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'singleFileUpload':
          return _react2.default.createElement(_UiSingleFileUpload2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'multiFileUpload':
          return _react2.default.createElement(_UiMultiFileUpload2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'pan':
          return _react2.default.createElement(_UiPanCard2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'aadhar':
          return _react2.default.createElement(_UiAadharCard2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'pinCode':
          return _react2.default.createElement(_UiPinCode2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'label':
          return _react2.default.createElement(_UiLabel2.default, { tabIndex: index, getVal: _this.props.getVal, item: item });
        case 'radio':
          return _react2.default.createElement(_UiRadioButton2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        case 'textSearch':
          return _react2.default.createElement(_UiTextSearch2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, autoComHandler: _this.props.autoComHandler });
        case 'documentList':
          return _react2.default.createElement(_UiDocumentList2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, readonly: screen === "view" ? "true" : "" });
        case 'arrayText':
        case 'arrayNumber':
          return _react2.default.createElement(_UiArrayField2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, readonly: screen === "view" ? "true" : "" });
        case 'fileTable':
          return _react2.default.createElement(_UiFileTable2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, readonly: screen === "view" ? "true" : screenView ? true : "" });
        case 'tableList':
          return _react2.default.createElement(_UiMultiFieldTable2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, screen: screen });
        case 'workflow':
          return _react2.default.createElement(_UiWorkflow2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, initiateWF: _this.props.initiateWF, workflowId: _this.props.workflowId || "" });
        case 'timePicker':
          return _react2.default.createElement(_UiTimeField2.default, { tabIndex: index, ui: _this.props.ui,
            getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors,
            handler: _this.props.handler, screen: screen });
        case 'window':
          return _react2.default.createElement(_UiWindowForm2.default, { tabIndex: index, ui: _this.props.ui,
            getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors,
            handler: _this.props.handler, screen: screen });
        case 'calendar':
          return _react2.default.createElement(_UiCalendar2.default, { ui: _this.props.ui, getVal: _this.props.getVal,
            item: item, fieldErrors: _this.props.fieldErrors,
            handler: _this.props.handler, screen: screen });
        case 'customComponent':
          // console.log(item.path);
          // {
          //   "name": "customReport",
          //   "jsonPath": "Connection.property.nameOfApplicant",
          //   "label": "wc.create.groups.applicantDetails.nameOfApplicant",
          //   "pattern": "",
          //   "type": "customComponent",
          //   "isRequired": false,
          //   "isDisabled": true,
          //   "requiredErrMsg": "",
          //   "patternErrMsg": "",
          //   "path":"./CustomComponent"
          // }

          var CustomComonent = require(item.path + ".js").default;
          return _react2.default.createElement(CustomComonent, null);
        case 'hyperLink':
          return _react2.default.createElement(_UiHyperLink2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });

        case 'googleMaps':
          return _react2.default.createElement(_UigoogleMaps2.default, { tabIndex: index, ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
        // return <div></div>
      }
    };

    _this.state = {};
    return _this;
  }

  _createClass(ShowFields, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          groups = _props.groups,
          noCols = _props.noCols,
          uiFramework = _props.uiFramework;

      return _react2.default.createElement(
        'div',
        null,
        this.renderGroups(groups, noCols, uiFramework)
      );
    }
  }]);

  return ShowFields;
}(_react.Component);

exports.default = ShowFields;
