// egov-ui-react-framework/components/UiArrayField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrap = require('react-bootstrap');

var _TextField = require('material-ui/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _FloatingActionButton = require('material-ui/FloatingActionButton');

var _FloatingActionButton2 = _interopRequireDefault(_FloatingActionButton);

var _Dialog = require('material-ui/Dialog');

var _Dialog2 = _interopRequireDefault(_Dialog);

var _FlatButton = require('material-ui/FlatButton');

var _FlatButton2 = _interopRequireDefault(_FlatButton);

var _egovCommonUtility = require('egov-common-utility');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiArrayField = function (_Component) {
	_inherits(UiArrayField, _Component);

	function UiArrayField(props) {
		_classCallCheck(this, UiArrayField);

		var _this = _possibleConstructorReturn(this, (UiArrayField.__proto__ || Object.getPrototypeOf(UiArrayField)).call(this, props));

		_this.renderField = function (item) {
			var val = _this.props.getVal(item.jsonPath);
			if (_this.props.readonly === "true") {
				return _react2.default.createElement(
					'div',
					null,
					_react2.default.createElement(
						_reactBootstrap.Col,
						{ xs: 12 },
						_react2.default.createElement(
							'label',
							null,
							_react2.default.createElement(
								'span',
								{ style: { "fontWeight": 500, "fontSize": "13px" } },
								(0, _egovCommonUtility.translate)(item.label)
							)
						)
					),
					_react2.default.createElement(
						_reactBootstrap.Col,
						{ xs: 12 },
						val && val.constructor == Array ? val.join(", ") : ""
					)
				);
			} else {
				return _react2.default.createElement(
					_reactBootstrap.Row,
					null,
					_react2.default.createElement(
						_reactBootstrap.Col,
						{ xs: 12, md: 6 },
						_react2.default.createElement(_TextField2.default, {
							className: 'cutustom-form-controll-for-textfield',
							id: item.jsonPath.split(".").join("-"),
							floatingLabelStyle: { "color": "#A9A9A9", "fontSize": "20px", "white-space": "nowrap" },
							inputStyle: { "color": "#5F5C57" },
							floatingLabelFixed: true,
							maxLength: item.maxLength || "",
							style: { "display": item.hide ? 'none' : 'inline-block' },
							errorStyle: { "float": "left" },
							fullWidth: true,
							floatingLabelText: _react2.default.createElement(
								'span',
								null,
								item.label,
								' ',
								_react2.default.createElement(
									'span',
									{ style: { "color": "#FF0000" } },
									item.isRequired ? " *" : ""
								)
							)
							//value = {this.state.valueList.join(", ")}
							, value: val && val.constructor == Array ? val.join(", ") : "",
							disabled: true
						})
					),
					_react2.default.createElement(
						_reactBootstrap.Col,
						{ xs: 12, md: 6 },
						_react2.default.createElement(
							_FloatingActionButton2.default,
							{
								style: { marginTop: 39 },
								mini: true,
								onClick: function onClick() {
									_this.handleOpen();
								} },
							_react2.default.createElement(
								'i',
								{ className: 'material-icons' },
								'add'
							)
						)
					)
				);
			}
		};

		_this.renderArrayField = function (item) {
			switch (_this.props.ui) {
				case 'google':
					return _react2.default.createElement(
						'div',
						null,
						_this.renderField(item),
						_react2.default.createElement(
							_Dialog2.default,
							{
								title: _this.props.item.label,
								actions: _react2.default.createElement(_FlatButton2.default, {
									label: (0, _egovCommonUtility.translate)("pt.create.button.viewdcb.close"),
									primary: true,
									onClick: _this.handleClose
								}),
								modal: false,
								open: _this.state.open,
								onRequestClose: _this.handleClose
							},
							_react2.default.createElement(
								_reactBootstrap.Row,
								null,
								_react2.default.createElement(
									_reactBootstrap.Col,
									{ xs: 12, md: 8 },
									_react2.default.createElement(_TextField2.default, {
										className: 'cutustom-form-controll-for-textfield',
										floatingLabelStyle: { "color": "#696969", "fontSize": "20px", "white-space": "nowrap" },
										inputStyle: { "color": "#5F5C57" },
										floatingLabelFixed: true,
										value: _this.state.currentValue,
										maxLength: _this.props.item.maxLength || "",
										errorStyle: { "float": "left" },
										fullWidth: true,
										floatingLabelText: _react2.default.createElement(
											'span',
											null,
											_this.props.item.label
										),
										errorText: _this.state.currentValue && _this.props.item.pattern && !new RegExp(_this.props.item.pattern).test(_this.state.currentValue) ? _this.props.item.patternErrMsg || "Invalid Pattern" : "",
										onChange: function onChange(e) {
											if (e.target.value) {
												e.target.value = e.target.value.replace(/^\s*/, "");
												if (e.target.value[e.target.value.length - 1] == " " && e.target.value[e.target.value.length - 2] == " ") return;
											}

											if (_this.props.item.type == "arrayNumber" && e.target.value && !/^\d.*$/.test(e.target.value)) return;
											_this.setState({ currentValue: e.target.value });
										} })
								),
								_react2.default.createElement(
									_reactBootstrap.Col,
									{ xs: 12, md: 4 },
									_react2.default.createElement(_FlatButton2.default, {
										label: (0, _egovCommonUtility.translate)("pt.create.groups.ownerDetails.fields.add"),
										disabled: !_this.state.currentValue || _this.state.currentValue && _this.props.item.pattern && !new RegExp(_this.props.item.pattern).test(_this.state.currentValue),
										secondary: true,
										style: { "marginTop": 39 },
										onClick: function onClick(e) {
											var list = [].concat(_toConsumableArray(_this.state.valueList));
											list.push(_this.state.currentValue);
											_this.setState({
												valueList: list,
												currentValue: ''
											}, function () {
												_this.props.handler({ target: { value: _this.state.valueList } }, _this.props.item.jsonPath, _this.props.item.isRequired ? true : false, '', _this.props.item.requiredErrMsg, _this.props.item.patternErrMsg);
											});
										}
									})
								)
							),
							_react2.default.createElement('br', null),
							_react2.default.createElement(
								_reactBootstrap.Table,
								{ className: 'table table-striped table-bordered', responsive: true },
								_react2.default.createElement(
									'thead',
									null,
									_react2.default.createElement(
										'tr',
										null,
										_react2.default.createElement(
											'th',
											null,
											'#'
										),
										_react2.default.createElement(
											'th',
											null,
											(0, _egovCommonUtility.translate)("collection.pay.value")
										),
										_react2.default.createElement(
											'th',
											null,
											(0, _egovCommonUtility.translate)("reports.common.action")
										)
									)
								),
								_react2.default.createElement(
									'tbody',
									null,
									_this.state.valueList.map(function (v, i) {
										return _react2.default.createElement(
											'tr',
											{ key: i },
											_react2.default.createElement(
												'td',
												null,
												i + 1
											),
											_react2.default.createElement(
												'td',
												null,
												v
											),
											_react2.default.createElement(
												'td',
												null,
												_react2.default.createElement(
													'div',
													{ className: 'material-icons', onClick: function onClick() {
															_this.valueFromList(i);
														} },
													'delete'
												)
											)
										);
									})
								)
							)
						)
					);
			}
		};

		_this.handleOpen = function () {
			_this.setState({
				open: true
			});
		};

		_this.handleClose = function () {
			_this.setState({
				open: false
			});
		};

		_this.valueFromList = function (index) {
			var list = [].concat(_toConsumableArray(_this.state.valueList));
			list.splice(index, 1);
			_this.setState({
				valueList: list
			}, function () {
				_this.props.handler({ target: { value: _this.state.valueList.length ? _this.state.valueList : "" } }, _this.props.item.jsonPath, _this.props.item.isRequired ? true : false, '', _this.props.item.requiredErrMsg, _this.props.item.patternErrMsg);
			});
		};

		_this.state = {
			open: false,
			currentValue: '',
			valueList: []
		};
		return _this;
	}

	_createClass(UiArrayField, [{
		key: 'componentWillReceiveProps',
		value: function componentWillReceiveProps(nextProps) {
			var arrayValue = this.props.getVal(this.props.item.jsonPath);
			var valueList = this.state.valueList;

			if (_lodash2.default.isArray(arrayValue) && JSON.stringify(arrayValue) != JSON.stringify(valueList)) {
				this.setState({
					valueList: arrayValue
				});
			}
		}
	}, {
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				null,
				this.renderArrayField(this.props.item)
			);
		}
	}]);

	return UiArrayField;
}(_react.Component);

exports.default = UiArrayField;
