// egov-ui-react-framework/components/UiAutoComplete.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _SelectField = require('material-ui/SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _AutoComplete = require('material-ui/AutoComplete');

var _AutoComplete2 = _interopRequireDefault(_AutoComplete);

var _MenuItem = require('material-ui/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _egovCommonUtility = require('egov-common-utility');

var _jsonpath = require('jsonpath');

var _jsonpath2 = _interopRequireDefault(_jsonpath);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var searchTextCom = "";

var UiAutoComplete = function (_Component) {
	_inherits(UiAutoComplete, _Component);

	function UiAutoComplete(props) {
		_classCallCheck(this, UiAutoComplete);

		var _this = _possibleConstructorReturn(this, (UiAutoComplete.__proto__ || Object.getPrototypeOf(UiAutoComplete)).call(this, props));

		_this.handleUpdateInput = function (searchText) {
			_this.setState({
				searchText: searchText
			});
		};

		_this.callAPI = function (keyUpValue) {
			var _this$props = _this.props,
			    item = _this$props.item,
			    setDropDownData = _this$props.setDropDownData,
			    useTimestamp = _this$props.useTimestamp;
			// console.log('API called:', item.hasOwnProperty("url"), item.url.search("\\|"), item.url.search("{"));

			if (item.hasOwnProperty("url") && item.url && item.url.search("\\|") > -1 && item.url.search("{") == -1) {
				//console.log(item.url.split("?"));
				var splitArray = item.url.split("?");
				var context = "";
				var id = {};
				// id[splitArray[1].split("&")[1].split("=")[0]]=e.target.value;
				for (var j = 0; j < splitArray[0].split("/").length; j++) {
					if (j == splitArray[0].split("/").length - 1) {
						context += splitArray[0].split("/")[j];
					} else {
						context += splitArray[0].split("/")[j] + "/";
					}
				}

				var queryStringObject = splitArray[1].split("|")[0].split("&");
				for (var i = 0; i < queryStringObject.length; i++) {
					//console.log(queryStringObject[i], queryStringObject[i].split("=")[0], queryStringObject[i].split("=")[1]);
					if (i) {
						if (keyUpValue) id[queryStringObject[i].split("=")[0]] = keyUpValue ? keyUpValue : queryStringObject[i].split("=")[1];else id[queryStringObject[i].split("=")[0]] = queryStringObject[i].split("=")[1];
					}
				}

				//console.log(id);

				var response = _egovCommonUtility.Api.commonApiPost(context, id, {}, "", useTimestamp || false).then(function (response) {

					if (response) {

						var queries = splitArray[1].split("|");
						var keys = _jsonpath2.default.query(response, queries[1]);
						var values = _jsonpath2.default.query(response, queries[2]);

						var others = [];
						if (queries.length > 3) {
							for (var _i = 3; _i < queries.length; _i++) {
								others.push(_jsonpath2.default.query(response, queries[_i]) || undefined);
							}
						}

						var dropDownData = [];
						for (var k = 0; k < keys.length; k++) {
							var obj = {};
							obj["key"] = keys[k] && keys[k].toString();
							obj["value"] = values[k];

							if (others && others.length > 0) {
								var otherItemDatas = [];
								for (var _i2 = 0; _i2 < others.length; _i2++) {
									otherItemDatas.push(others[_i2][k] || undefined);
								}
								obj['others'] = otherItemDatas;
							}

							if (item.hasOwnProperty("isKeyValuePair") && item.isKeyValuePair) {
								obj["value"] = keys[k] + "-" + values[k];
							}
							dropDownData.push(obj);
						}
						setDropDownData(item.jsonPath, dropDownData);
					}
				}, function (err) {
					console.log(err);
				});
			} else if (item.hasOwnProperty("defaultValue") && _typeof(item.defaultValue) == "object") {
				setDropDownData(item.jsonPath, item.defaultValue);
			}
		};

		_this.renderAutoComplete = function (item) {
			var dropDownData = _this.props.dropDownData;

			var dataSourceConfig = {
				text: 'value',
				value: 'key'
			};

			// console.log(dropDownData.hasOwnProperty(item.jsonpath) && dropDownData[item.jsonpath].replace(".", "\."));
			// console.log(dropDownData);
			// console.log(dropDownData.hasOwnProperty(item.jsonPath));
			// console.log(searchTextCom);
			switch (_this.props.ui) {
				case 'google':
					// let {dropDownData}=this.state;
					return _react2.default.createElement(
						'div',
						null,
						_react2.default.createElement(_AutoComplete2.default, {
							className: 'custom-form-control-for-textfield',
							id: item.jsonPath.split(".").join("-"),
							listStyle: { maxHeight: 100, overflow: 'auto' },
							onUpdateInput: _this.handleUpdateInput,
							filter: function filter(searchText, key) {
								return key.toLowerCase().includes(searchText.toLowerCase());
							},
							floatingLabelStyle: { "color": item.isDisabled ? "#A9A9A9" : "#696969", "fontSize": "20px", "white-space": "nowrap" },
							inputStyle: { "color": "#5F5C57" },
							floatingLabelFixed: true,
							style: { "display": item.hide ? 'none' : 'inline-block' },
							errorStyle: { "float": "left" },
							dataSource: dropDownData.hasOwnProperty(item.jsonPath) ? dropDownData[item.jsonPath] : [],
							dataSourceConfig: dataSourceConfig,
							floatingLabelText: _react2.default.createElement(
								'span',
								null,
								item.label,
								' ',
								_react2.default.createElement(
									'span',
									{ style: { "color": "#FF0000" } },
									item.isRequired ? " *" : ""
								)
							),
							fullWidth: true,
							searchText: _this.state.searchText || searchTextCom,
							disabled: item.isDisabled,
							errorText: _this.props.fieldErrors[item.jsonPath],
							onKeyUp: function onKeyUp(e) {
								item.onLoad == false ? _this.callAPI(e.target.value) : '';
								//this.props.handler({target: {value: (item.allowWrite ? e.target.value : "")}}, item.jsonPath, item.isRequired ? true : false, '', item.requiredErrMsg, item.patternErrMsg)
							},
							onNewRequest: function onNewRequest(value, index) {
								_this.props.handler({ target: { value: value.key } }, item.jsonPath, item.isRequired ? true : false, '', item.requiredErrMsg, item.patternErrMsg);
								_this.handleUpdateInput(value.value);
								if (_this.props.autoComHandler && item.autoCompleteDependancy) {
									_this.props.autoComHandler(item.autoCompleteDependancy, item.jsonPath);
								}
							}
						})
					);
			}
		};

		_this.state = {
			searchText: ""
		};
		return _this;
	}

	_createClass(UiAutoComplete, [{
		key: 'initData',
		value: function initData() {
			var _props = this.props,
			    item = _props.item,
			    setDropDownData = _props.setDropDownData;

			if (item.onLoad == false) {
				setDropDownData(item.jsonPath, []);
			} else {
				this.callAPI('');
			}
		}

		// componentWillReceiveProps(nextProps, nextState) {
		//  		if(!_.isEqual(nextProps, this.props)) {
		//  			this.initData(nextProps);
		//  		}
		//  	}

	}, {
		key: 'componentDidMount',
		value: function componentDidMount() {
			this.initData();
		}
	}, {
		key: 'render',
		value: function render() {
			var self = this;
			searchTextCom = self.props.getVal(this.props.item.jsonPath);
			return _react2.default.createElement(
				'div',
				null,
				this.renderAutoComplete(this.props.item)
			);
		}
	}]);

	return UiAutoComplete;
}(_react.Component);

var mapStateToProps = function mapStateToProps(state) {
	return { dropDownData: state.framework.dropDownData };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
	return {
		setDropDownData: function setDropDownData(fieldName, dropDownData) {
			dispatch({ type: "SET_DROPDWON_DATA", fieldName: fieldName, dropDownData: dropDownData });
		}
	};
};
exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(UiAutoComplete);
