// egov-ui-react-framework/components/UiBackButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _reactRouter = require('react-router');

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiBackButton = function (_Component) {
	_inherits(UiBackButton, _Component);

	function UiBackButton(props) {
		_classCallCheck(this, UiBackButton);

		var _this = _possibleConstructorReturn(this, (UiBackButton.__proto__ || Object.getPrototypeOf(UiBackButton)).call(this, props));

		_this.back = function (e) {
			var setRoute = _this.props.setRoute;

			setRoute("/" + localStorage.getItem("returnUrl"));
		};

		return _this;
	}

	_createClass(UiBackButton, [{
		key: 'render',
		value: function render() {
			var back = this.back;

			return _react2.default.createElement(
				'div',
				{ style: { "textAlign": "right",
						"paddingRight": "15px" } },
				_react2.default.createElement('br', null),
				_react2.default.createElement(_RaisedButton2.default, { type: 'button', onClick: function onClick(e) {
						back(e);
					}, primary: true, label: "Back" }),
				_react2.default.createElement('br', null),
				_react2.default.createElement('br', null)
			);
		}
	}]);

	return UiBackButton;
}(_react.Component);

var mapStateToProps = function mapStateToProps(state) {
	return {};
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
	return {
		setRoute: function setRoute(route) {
			return dispatch({ type: "SET_ROUTE", route: route });
		}
	};
};

exports.default = (0, _reactRouter.withRouter)((0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(UiBackButton));
