// egov-ui-react-framework/components/UiCalendar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _egovCommonUtility = require('egov-common-utility');

var _reactBigCalendar = require('react-big-calendar');

var _reactBigCalendar2 = _interopRequireDefault(_reactBigCalendar);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _reactRedux = require('react-redux');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

_reactBigCalendar2.default.momentLocalizer(_moment2.default);

var UiCalendar = function (_Component) {
    _inherits(UiCalendar, _Component);

    function UiCalendar(props) {
        _classCallCheck(this, UiCalendar);

        var _this = _possibleConstructorReturn(this, (UiCalendar.__proto__ || Object.getPrototypeOf(UiCalendar)).call(this, props));

        _this.renderCalendar = function (item) {
            var allViews = Object.keys(_reactBigCalendar2.default.Views).map(function (k) {
                return _reactBigCalendar2.default.Views[k];
            });
            switch (_this.props.ui) {
                case 'google':
                    return _react2.default.createElement(
                        'div',
                        null,
                        _react2.default.createElement(_reactBigCalendar2.default, {
                            events: _this.getValue(item.jsonPath),
                            views: allViews,
                            step: 60,
                            defaultDate: new Date(),
                            components: {
                                event: _this.Event,
                                agenda: {
                                    event: _this.EventAgenda
                                }
                            }
                        })
                    );
            }
        };

        return _this;
    }

    _createClass(UiCalendar, [{
        key: 'Event',
        value: function Event(_ref) {
            var event = _ref.event;

            return _react2.default.createElement(
                'span',
                null,
                _react2.default.createElement(
                    'strong',
                    null,
                    event.title
                ),
                event.desc && ':  ' + event.desc
            );
        }
    }, {
        key: 'EventAgenda',
        value: function EventAgenda(_ref2) {
            var event = _ref2.event;

            return _react2.default.createElement(
                'span',
                null,
                _react2.default.createElement(
                    'em',
                    { style: { color: 'blue' } },
                    event.title
                ),
                _react2.default.createElement(
                    'p',
                    null,
                    event.desc
                )
            );
        }
    }, {
        key: 'getValue',
        value: function getValue(item) {
            var eventsObj = [];
            var edata = this.props.getVal(item);
            if (edata && edata.length > 0) {
                edata.map(function (v, i) {
                    var hearingTimeDate = new Date();
                    if (v.nextHearingDate && v.nextHearingTime) {
                        var timeDate = new Date(parseInt(v.nextHearingTime));
                        var dataDate = new Date(v.nextHearingDate);
                        hearingTimeDate = timeDate.setDate(dataDate.getDate());
                    } else {
                        hearingTimeDate = v.nextHearingDate ? v.nextHearingDate : new Date();
                    }
                    eventsObj.push({
                        'title': 'Case' + v.caseNo,
                        'start': new Date(hearingTimeDate),
                        'end': new Date(hearingTimeDate),
                        'desc': v.departmentConcernPerson
                    });
                });
            }

            return eventsObj;
        }
    }, {
        key: 'render',
        value: function render() {
            return this.renderCalendar(this.props.item);
        }
    }]);

    return UiCalendar;
}(_react.Component);

exports.default = UiCalendar;
