// egov-ui-react-framework/components/UiDocumentList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TextField = require('material-ui/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _egovCommonUtility = require('egov-common-utility');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _reactBootstrap = require('react-bootstrap');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiDocumentList = function (_Component) {
	_inherits(UiDocumentList, _Component);

	function UiDocumentList(props) {
		_classCallCheck(this, UiDocumentList);

		var _this = _possibleConstructorReturn(this, (UiDocumentList.__proto__ || Object.getPrototypeOf(UiDocumentList)).call(this, props));

		_this.renderFileObject = function (item, i) {
			var self = _this;
			if (self.props.readonly) {
				return _react2.default.createElement(
					'a',
					{ href: window.location.origin + "/filestore/v1/files/id?tenantId=" + localStorage.tenantId + "&fileStoreId=" + self.props.getVal(item.jsonPath + "[" + i + "].fileStoreId"), target: '_blank' },
					(0, _egovCommonUtility.translate)("wc.craete.file.Download")
				);
			} else {
				return _react2.default.createElement('input', { type: 'file', onChange: function onChange(e) {
						self.props.handler({ target: { value: e.target.files[0] } }, item.jsonPath + "[" + i + "].fileStoreId", true, '', item.requiredErrMsg, item.patternErrMsg);
						self.props.handler({ target: { value: e.target.files[0].name } }, item.jsonPath + "[" + i + "].name", true, '', item.requiredErrMsg, item.patternErrMsg);
					} });
			}
		};

		_this.renderDocumentList = function (item) {
			var self = _this;
			switch (_this.props.ui) {
				case 'google':
					/*return self.state.documents.map(function(doc, i) {
     	return (
     		<Col xs={12} md={12}>
     			<TextField
     				id={item.jsonPath.split(".").join("-")}
     				floatingLabelStyle={{"color": "#696969", "fontSize": "20px"}}
     				floatingLabelFixed={true}
     				fullWidth={true}
     				type="text"
     				floatingLabelText={<span>{item.label} <span style={{"color": "#FF0000"}}>{item.isRequired ? " *" : ""}</span></span>}
     				value={self.props.getVal(item.jsonPath + "[" + i + "].name")}
     				disabled={item.isDisabled}
     				errorText={self.props.fieldErrors[item.jsonPath]}
     				onChange={(e) => self.props.handler(e, (item.jsonPath + "[" + i + "].name"), true, '', '', '')} />
     			<RaisedButton
     			  containerElement='label'
     			  fullWidth={true}
     			  value={self.props.getVal(item.jsonPath + "[" + i + "].fileStoreId")}
     			  label={doc["displayName"]}>
     			    <input type="file" style={{ display: 'none' }} onChange={(e) => {
     			    	self.props.handler({target:{value: e.target.files[0]}}, (item.jsonPath + "[" + i + "].fileStoreId"), true, '', item.requiredErrMsg, item.patternErrMsg)
     			    	self.props.handler({target:{value: e.target.files[0].name}}, (item.jsonPath + "[" + i + "].name"), true, '', item.requiredErrMsg, item.patternErrMsg)
     			    }}/>
     			</RaisedButton>
     		</Col>
     	)
     })*/
					return _react2.default.createElement(
						_reactBootstrap.Table,
						{ className: 'table table-striped table-bordered', cellspacing: '0', width: '100%', responsive: true },
						_react2.default.createElement(
							'thead',
							null,
							_react2.default.createElement(
								'th',
								null,
								'#'
							),
							_react2.default.createElement(
								'th',
								null,
								(0, _egovCommonUtility.translate)("reports.common.businesskey")
							),
							_react2.default.createElement(
								'th',
								null,
								(0, _egovCommonUtility.translate)("wc.create.name")
							),
							_react2.default.createElement(
								'th',
								null,
								(0, _egovCommonUtility.translate)("tl.create.license.table.file")
							)
						),
						_react2.default.createElement(
							'tbody',
							null,
							self.state.documents.map(function (doc, i) {
								return _react2.default.createElement(
									'tr',
									null,
									_react2.default.createElement(
										'td',
										null,
										i + 1
									),
									_react2.default.createElement(
										'td',
										null,
										doc["displayName"]
									),
									_react2.default.createElement(
										'td',
										null,
										_react2.default.createElement(_TextField2.default, {
											fullWidth: true,
											disabled: self.props.readonly,
											type: 'text',
											value: self.props.getVal(item.jsonPath + "[" + i + "].name"),
											errorText: self.props.fieldErrors[item.jsonPath],
											onChange: function onChange(e) {
												return self.props.handler(e, item.jsonPath + "[" + i + "].name", true, '', '', '');
											} })
									),
									_react2.default.createElement(
										'td',
										null,
										self.renderFileObject(item, i)
									)
								);
							})
						)
					);
			}
		};

		_this.state = {
			documents: []
		};
		return _this;
	}

	_createClass(UiDocumentList, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var self = this;
			var _props = this.props,
			    item = _props.item,
			    useTimestamp = _props.useTimestamp,
			    handler = _props.handler;

			if (item.url) {
				var context = item.url.split("?")[0];
				var query = {};
				if (item.url.split("?")[1]) {
					var queryArr = item.url.split("?")[1].split("&");
					for (var i = 0; i < queryArr.length; i++) {
						query[queryArr[i].split("=")[0]] = queryArr[i].split("=")[1];
					}
				}

				_egovCommonUtility.Api.commonApiPost(context, query, {}, "", useTimestamp, false, null, item.appendUserInfo ? JSON.parse(localStorage.userRequest) : null).then(function (res) {
					var documents = [];
					var arr = _lodash2.default.get(res, item.pathToArray);
					if (arr && arr.length) {
						for (var k = 0; k < arr.length; k++) {
							var temp = {
								"fileStoreId": "",
								"displayName": arr[k][item.displayNameJsonPath],
								"name": ""
							};
							for (var i = 0; i < item.autoFillFields.length; i++) {
								temp[item.autoFillFields[i].name] = arr[k][item.autoFillFields[i].jsonPath];
							}

							documents.push(temp);
						}

						self.setState({ documents: documents }, function () {
							self.props.handler({ target: { value: documents } }, item.jsonPath, false, '');
						});
					}
				}, function (err) {
					console.log(err);
				});
			}
		}
	}, {
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				null,
				this.renderDocumentList(this.props.item)
			);
		}
	}]);

	return UiDocumentList;
}(_react.Component);

exports.default = UiDocumentList;
