// egov-ui-react-framework/components/UiDynamicTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _egovCommonUtility = require('egov-common-utility');

var _reactBootstrap = require('react-bootstrap');

var _Card = require('material-ui/Card');

var _UiTextField = require('./UiTextField');

var _UiTextField2 = _interopRequireDefault(_UiTextField);

var _UiSelectField = require('./UiSelectField');

var _UiSelectField2 = _interopRequireDefault(_UiSelectField);

var _UiButton = require('./UiButton');

var _UiButton2 = _interopRequireDefault(_UiButton);

var _UiCheckBox = require('./UiCheckBox');

var _UiCheckBox2 = _interopRequireDefault(_UiCheckBox);

var _UiEmailField = require('./UiEmailField');

var _UiEmailField2 = _interopRequireDefault(_UiEmailField);

var _UiMobileNumber = require('./UiMobileNumber');

var _UiMobileNumber2 = _interopRequireDefault(_UiMobileNumber);

var _UiTextArea = require('./UiTextArea');

var _UiTextArea2 = _interopRequireDefault(_UiTextArea);

var _UiMultiSelectField = require('./UiMultiSelectField');

var _UiMultiSelectField2 = _interopRequireDefault(_UiMultiSelectField);

var _UiNumberField = require('./UiNumberField');

var _UiNumberField2 = _interopRequireDefault(_UiNumberField);

var _UiDatePicker = require('./UiDatePicker');

var _UiDatePicker2 = _interopRequireDefault(_UiDatePicker);

var _UiMultiFileUpload = require('./UiMultiFileUpload');

var _UiMultiFileUpload2 = _interopRequireDefault(_UiMultiFileUpload);

var _UiSingleFileUpload = require('./UiSingleFileUpload');

var _UiSingleFileUpload2 = _interopRequireDefault(_UiSingleFileUpload);

var _UiAadharCard = require('./UiAadharCard');

var _UiAadharCard2 = _interopRequireDefault(_UiAadharCard);

var _UiPanCard = require('./UiPanCard');

var _UiPanCard2 = _interopRequireDefault(_UiPanCard);

var _UiLabel = require('./UiLabel');

var _UiLabel2 = _interopRequireDefault(_UiLabel);

var _UiRadioButton = require('./UiRadioButton');

var _UiRadioButton2 = _interopRequireDefault(_UiRadioButton);

var _UiTextSearch = require('./UiTextSearch');

var _UiTextSearch2 = _interopRequireDefault(_UiTextSearch);

var _UiDocumentList = require('./UiDocumentList');

var _UiDocumentList2 = _interopRequireDefault(_UiDocumentList);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var $ = require('jquery');
$.DataTable = require('datatables.net');
var dt = require('datatables.net-bs');

var UiDynamicTable = function (_Component) {
	_inherits(UiDynamicTable, _Component);

	function UiDynamicTable(props) {
		_classCallCheck(this, UiDynamicTable);

		var _this = _possibleConstructorReturn(this, (UiDynamicTable.__proto__ || Object.getPrototypeOf(UiDynamicTable)).call(this, props));

		_this.renderFields = function (item) {
			switch (item.type) {
				case 'text':
					return _react2.default.createElement(_UiTextField2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'singleValueList':
					return _react2.default.createElement(_UiSelectField2.default, { ui: _this.props.ui, useTimestamp: _this.props.useTimestamp, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'multiValueList':
					return _react2.default.createElement(_UiSingleFileUpload2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'number':
					return _react2.default.createElement(_UiNumberField2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'textarea':
					return _react2.default.createElement(_UiTextArea2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'mobileNumber':
					return _react2.default.createElement(_UiMobileNumber2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'checkbox':
					return _react2.default.createElement(_UiCheckBox2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'email':
					return _react2.default.createElement(_UiEmailField2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'button':
					return _react2.default.createElement(_UiButton2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'datePicker':
					return _react2.default.createElement(_UiDatePicker2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;
				case 'singleFileUpload':
					return _react2.default.createElement(_UiSingleFileUpload2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
					break;

				case 'multiFileUpload':
					return _react2.default.createElement(_UiMultiSelectField2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
				case 'pan':
					return _react2.default.createElement(_UiPanCard2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
				case 'aadhar':
					return _react2.default.createElement(_UiAadharCard2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
				case 'label':
					return _react2.default.createElement(_UiLabel2.default, { getVal: _this.props.getVal, item: item });
				case 'radio':
					return _react2.default.createElement(_UiRadioButton2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
				case 'textSearch':
					return _react2.default.createElement(_UiTextSearch2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler, autoComHandler: _this.props.autoComHandler });
				case 'documentList':
					return _react2.default.createElement(_UiDocumentList2.default, { ui: _this.props.ui, getVal: _this.props.getVal, item: item, fieldErrors: _this.props.fieldErrors, handler: _this.props.handler });
			}
		};

		return _this;
	}

	// 	componentWillMount() {
	//     $('#dynamicTable').DataTable({
	//        dom: 'lBfrtip',
	//        buttons: [],
	//        bDestroy: true,
	//        language: {
	//            "emptyTable": "No Records"
	//        }
	//     });
	//  	}
	//
	//  	componentWillUnmount() {
	//   	$('#dynamicTable').DataTable().destroy(true);
	// 	}
	//
	// 	componentWillUpdate() {
	// 		let {flag} = this.props;
	//     if(flag == 1) {
	//       flag = 0;
	//       $('#dynamicTable').dataTable().fnDestroy();
	//     }
	// }
	//
	// componentDidUpdate() {
	//     $('#dynamicTable').DataTable({
	//          dom: 'lBfrtip',
	//          buttons: [],
	//           ordering: false,
	//           bDestroy: true,
	//           language: {
	//              "emptyTable": "No Records"
	//           }
	//     });
	// 	}

	_createClass(UiDynamicTable, [{
		key: 'render',
		value: function render() {
			var self = this;
			var _props = this.props,
			    resultList = _props.resultList,
			    getVal = _props.getVal;
			var renderFields = this.renderFields;
			// console.log(resultList);

			// const totalAmount=()=>{
			// 	 let bill=getVal("Receipt[0].Bill");
			// 	 let sum=0;
			// 	 for (var i = 0; i < bill.length; i++) {
			// 		 	if (bill[i].hasOwnProperty("amountPaid")) {
			// 		 		sum+=bill[i].amountPaid;
			// 		 	}
			// 	 }
			// 	 return sum;
			// }

			var renderTable = function renderTable() {
				return _react2.default.createElement(
					_Card.Card,
					{ className: 'uiCard' },
					_react2.default.createElement(_Card.CardHeader, { title: _react2.default.createElement(
							'strong',
							null,
							' ',
							(0, _egovCommonUtility.translate)("ui.table.title"),
							' '
						) }),
					_react2.default.createElement(
						_Card.CardText,
						null,
						_react2.default.createElement(
							_reactBootstrap.Table,
							{ id: 'dynamicTable', bordered: true, responsive: true, className: 'table-striped' },
							_react2.default.createElement(
								'thead',
								null,
								_react2.default.createElement(
									'tr',
									null,
									resultList.resultHeader && resultList.resultHeader.length && resultList.resultHeader.map(function (item, i) {
										return _react2.default.createElement(
											'th',
											{ style: { "textAlign": "center" }, key: i },
											(0, _egovCommonUtility.translate)(item.label)
										);
									})
								)
							),
							_react2.default.createElement(
								'tbody',
								null,
								resultList.hasOwnProperty("resultValues") && resultList.resultValues.map(function (item, i) {
									return _react2.default.createElement(
										'tr',
										{ key: i },
										item.map(function (item2, i2) {
											return _react2.default.createElement(
												'td',
												{ style: { "textAlign": "center" }, key: i2 },
												renderFields(item2)
											);
										})
									);
								})
							)
						),
						_react2.default.createElement(
							'strong',
							null,
							"Amount paid (Rs)" + " " + getVal("Receipt[0].instrument.amount")
						)
					)
				);
			};

			return _react2.default.createElement(
				'div',
				null,
				resultList && renderTable()
			);
		}
	}]);

	return UiDynamicTable;
}(_react.Component);

var mapStateToProps = function mapStateToProps(state) {
	return { flag: state.report.flag };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
	return {
		setFlag: function setFlag(flag) {
			dispatch({ type: "SET_FLAG", flag: flag });
		}
	};
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(UiDynamicTable);
