// egov-ui-react-framework/components/UiFileTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrap = require('react-bootstrap');

var _TextField = require('material-ui/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _egovCommonUtility = require('egov-common-utility');

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _reactRedux = require('react-redux');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiFileTable = function (_Component) {
	_inherits(UiFileTable, _Component);

	function UiFileTable(props) {
		_classCallCheck(this, UiFileTable);

		var _this = _possibleConstructorReturn(this, (UiFileTable.__proto__ || Object.getPrototypeOf(UiFileTable)).call(this, props));

		_this.renderFileObject = function (item, i, idValue) {
			if (_this.props.readonly) {
				return _react2.default.createElement(
					'a',
					{ href: window.location.origin + "/filestore/v1/files/id?tenantId=" + localStorage.tenantId + "&fileStoreId=" + _this.props.getVal(item.jsonPath + "[" + i + "]." + item.fileList.id), target: '_blank' },
					(0, _egovCommonUtility.translate)("wc.craete.file.Download")
				);
			} else {
				return _react2.default.createElement(
					'div',
					{ date: idValue },
					_react2.default.createElement('input', { type: 'file', onChange: function onChange(e) {
							_this.props.handler({ target: { value: e.target.files[0] } }, item.jsonPath + "[" + i + "]." + item.fileList.id, item.isRequired, '', item.requiredErrMsg, item.patternErrMsg);
						} })
				);
			}
		};

		_this.renderRowList = function (item) {
			var fileRowCount = _this.state.fileRowCount;


			var arr = [].concat(_toConsumableArray(Array(fileRowCount).keys()));
			var value = "";

			return _react2.default.createElement(
				'tbody',
				null,
				arr && arr.length && arr.map(function (v, i) {
					var fileListName = item.jsonPath + "[" + i + "]." + item.fileList.name;
					var fileStoreId = item.jsonPath + "[" + i + "]." + item.fileList.id;

					value = _this.props.getVal(fileListName);
					var idValue = _this.props.getVal(fileStoreId);
					if (value || idValue || !_this.props.readonly) {
						return _react2.default.createElement(
							'tr',
							{ key: i },
							_react2.default.createElement(
								'td',
								null,
								i + 1
							),
							_react2.default.createElement(
								'td',
								null,
								_react2.default.createElement(_TextField2.default, {
									className: 'cutustom-form-controll-for-textfield',
									id: item.jsonPath.split(".").join("-"),
									inputStyle: { "color": "#5F5C57" },
									errorStyle: { "float": "left" },
									fullWidth: true,
									disabled: _this.props.readonly,
									value: value,
									onChange: function onChange(e) {
										_this.props.handler(e, fileListName, item.isRequired, '', item.requiredErrMsg, item.patternErrMsg);
									} })
							),
							_react2.default.createElement(
								'td',
								null,
								_this.renderFileObject(item, i, idValue)
							),
							_this.props.readonly ? "" : _react2.default.createElement(
								'td',
								null,
								_react2.default.createElement(
									'div',
									{ className: 'material-icons', onClick: function onClick() {
											_this.deleteRow(i);
										} },
									'delete'
								)
							)
						);
					}
				})
			);
		};

		_this.renderFileTable = function (item) {
			switch (_this.props.ui) {
				case 'google':
					return _react2.default.createElement(
						'div',
						null,
						_react2.default.createElement(
							_reactBootstrap.Table,
							{ className: 'table table-striped table-bordered', responsive: true },
							_react2.default.createElement(
								'thead',
								null,
								_react2.default.createElement(
									'tr',
									null,
									_react2.default.createElement(
										'th',
										null,
										'#'
									),
									_react2.default.createElement(
										'th',
										null,
										(0, _egovCommonUtility.translate)("tl.create.license.table.documentName")
									),
									_react2.default.createElement(
										'th',
										null,
										(0, _egovCommonUtility.translate)("wc.create.groups.fileDetails.title")
									),
									_this.props.readonly ? "" : _react2.default.createElement(
										'th',
										null,
										(0, _egovCommonUtility.translate)("reports.common.action")
									)
								)
							),
							_this.renderRowList(item)
						),
						_this.props.readonly ? "" : _react2.default.createElement(
							'div',
							{ style: { "textAlign": "right" } },
							_react2.default.createElement(_RaisedButton2.default, { label: (0, _egovCommonUtility.translate)("pgr.lbl.add"), primary: true, onClick: function onClick() {
									_this.addNewRow();
								} })
						)
					);
			}
		};

		_this.deleteRow = function (ind) {
			var item = _this.props.item;
			//let formData = JSON.parse(JSON.stringify(this.props.formData));

			var formDataArray = _lodash2.default.get(_this.props.formData, item.jsonPath);
			if (formDataArray && formDataArray.length) {
				formDataArray.splice(ind, 1);
				var newArray = _lodash2.default.cloneDeep(formDataArray);
				var newFormData = _lodash2.default.set(_this.props.formData, item.jsonPath, newArray);
				_this.props.setFormData(newFormData);
			}

			_this.setState({
				fileRowCount: _this.state.fileRowCount - 1
			});
		};

		_this.state = {
			fileRowCount: 1,
			isintialLoad: false
		};
		return _this;
	}

	_createClass(UiFileTable, [{
		key: 'componentWillMount',
		value: function componentWillMount() {
			this.setState({
				fileRowCount: this.props.item.fileCount
			});
		}
	}, {
		key: 'componentWillReceiveProps',
		value: function componentWillReceiveProps(nextProps) {
			var _state = this.state,
			    fileRowCount = _state.fileRowCount,
			    isintialLoad = _state.isintialLoad;
			var item = nextProps.item;

			var rowCount = fileRowCount;
			if (!isintialLoad && item.fileCount && item.fileCount > fileRowCount) {
				this.setState({
					fileRowCount: item.fileCount,
					isintialLoad: true
				});
			} else if (!isintialLoad && !item.fileCount) {
				this.setState({
					fileRowCount: 10,
					isintialLoad: true
				});
			}
		}
	}, {
		key: 'addNewRow',
		value: function addNewRow() {
			this.setState({
				fileRowCount: this.state.fileRowCount + 1
			});
		}
	}, {
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				null,
				this.renderFileTable(this.props.item)
			);
		}
	}]);

	return UiFileTable;
}(_react.Component);

var mapStateToProps = function mapStateToProps(state) {
	return {
		formData: state.frameworkForm.form
	};
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
	return {
		setFormData: function setFormData(data) {
			dispatch({ type: "SET_FORM_DATA", data: data });
		}
	};
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(UiFileTable);
