// egov-ui-react-framework/components/UiLabel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrap = require('react-bootstrap');

var _egovCommonUtility = require('egov-common-utility');

var _jsonpath = require('jsonpath');

var _jsonpath2 = _interopRequireDefault(_jsonpath);

var _FlatButton = require('material-ui/FlatButton');

var _FlatButton2 = _interopRequireDefault(_FlatButton);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiLabel = function (_Component) {
  _inherits(UiLabel, _Component);

  function UiLabel(props) {
    _classCallCheck(this, UiLabel);

    var _this = _possibleConstructorReturn(this, (UiLabel.__proto__ || Object.getPrototypeOf(UiLabel)).call(this, props));

    _this.setVal = function () {
      var _this$props = _this.props,
          item = _this$props.item,
          useTimestamp = _this$props.useTimestamp;

      var self = _this;
      var val = _this.props.getVal(item.jsonPath, item.isDate);
      if (item.configUrl && item.url) {
        var _url = item.configUrl.split("?")[0];
        var qString = {};
        if (item.configUrl.split("?")[1]) {
          var qKeys = item.configUrl.split("?")[1].split("&");
          for (var _i = 0; _i < qKeys.length; _i++) {
            qString[qKeys[_i].split("=")[0]] = qKeys[_i].split("=")[1];
          }
        }

        _egovCommonUtility.Api.commonApiPost(_url, qString).then(function (config) {
          _url = item.url.split("?")[0];
          qString = {};
          var qKeys = item.url.split("?")[1].split("|")[0].split("&");
          for (var i = 0; i < qKeys.length; i++) {
            if (i) {
              if (/\}/.test(qKeys[i].split("=")[1])) qString[qKeys[i].split("=")[0]] = _lodash2.default.get(config, qKeys[i].split("=")[1].split("{")[1].split("}")[0]);else qString[qKeys[i].split("=")[0]] = qKeys[i].split("=")[1];
            }
          }

          _egovCommonUtility.Api.commonApiPost(_url, qString, {}, "", useTimestamp || false).then(function (response) {
            if (response) {
              var keys = _jsonpath2.default.query(response, item.url.split("?")[1].split("|")[1]);
              var values = _jsonpath2.default.query(response, item.url.split("?")[1].split("|")[2]);
              var dropDownData = [];
              for (var k = 0; k < keys.length; k++) {
                if (val == keys[k]) {
                  return self.setState({
                    value: values[k]
                  });
                }
              }
            }
          }, function (err) {
            console.log(err);
          });
        }).catch(function (err) {
          console.log(err);
        });
      }
      /*if (item.isConfig) {
         //=======================BASED ON APP CONFIG==========================//
         Api.commonApiPost('/wcms/masters/waterchargesconfig/_search', {
           name: "HIERACHYTYPEFORWC"
         }).then((res1) => {
           if(res1.WaterConfigurationValue && res1.WaterConfigurationValue[0] && res1.WaterConfigurationValue[0].value && res1.WaterConfigurationValue[0].value) {
              Api.commonApiPost('egov-location/boundarys/boundariesByBndryTypeNameAndHierarchyTypeName', {boundaryTypeName:"ZONE", hierarchyTypeName:res1.WaterConfigurationValue[0].value}).then((response)=>{
               if(response) {
      					let keys = jp.query(response,"$.Boundary.*.code");
                let values = jp.query(response,"$.Boundary.*.name");
                let dropDownData = [];
                for (var k = 0; k < keys.length; k++) {
                    if(val == keys[k]) {
                      return self.setState({
                        value: values[k]
                      })
                    }
                }
      				}
             }).catch((err)=> {
               console.log(err)
             })
           }
          }).catch((err) => {
             console.log(err);
         })
      	} */else if (val && item.hasOwnProperty("url") && item.url.search("\\|") > -1) {
          var splitArray = item.url.split("?");
          var context = "";
          var id = {};
          for (var j = 0; j < splitArray[0].split("/").length; j++) {
            context += splitArray[0].split("/")[j] + "/";
          }

          var queryStringObject = splitArray[1].split("|")[0].split("&");
          for (var i = 0; i < queryStringObject.length; i++) {
            if (i) {
              id[queryStringObject[i].split("=")[0]] = queryStringObject[i].split("=")[1];
            }
          }

          _egovCommonUtility.Api.commonApiPost(context, id, {}, "", useTimestamp || false).then(function (response) {
            if (response) {
              var keys = _jsonpath2.default.query(response, splitArray[1].split("|")[1]);
              var values = _jsonpath2.default.query(response, splitArray[1].split("|")[2]);
              var dropDownData = [];
              for (var k = 0; k < keys.length; k++) {
                if (val == keys[k]) {
                  return self.setState({
                    value: values[k]
                  });
                }
              }
            }
          }, function (err) {
            console.log(err);
          });
        } else if (val) {
          _this.setState({
            value: val
          });
        }
    };

    _this.renderLabel = function (item) {

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _reactBootstrap.Row,
          null,
          !item.hasOwnProperty("isLabel") ? _react2.default.createElement(
            _reactBootstrap.Col,
            { id: item.jsonPath.split(".").join("-"), style: item.hasOwnProperty("textAlign") ? { textAlign: item.textAlign } : { textAlign: "left" }, xs: 12 },
            _react2.default.createElement(
              'label',
              null,
              _react2.default.createElement(
                'span',
                { style: { "fontWeight": 600, "fontSize": "13px" } },
                (0, _egovCommonUtility.translate)(item.label)
              )
            )
          ) : "",
          item.hyperLink && (_this.state.value || _this.props.getVal(item.jsonPath, item.isDate)) ? _react2.default.createElement(
            _reactBootstrap.Col,
            { style: { textAlign: "left" }, xs: 12 },
            _react2.default.createElement(_FlatButton2.default, { label: _this.state.value || _this.props.getVal(item.jsonPath, item.isDate), primary: true })
          ) : _react2.default.createElement(
            _reactBootstrap.Col,
            { id: item.jsonPath.split(".").join("-"), style: item.hasOwnProperty("textAlign") ? { textAlign: item.textAlign } : { textAlign: "left" }, xs: 12 },
            _this.state.value || _this.props.getVal(item.jsonPath, item.isDate) || "NA"
          )
        ),
        _react2.default.createElement('br', null)
      );
    };

    _this.state = {
      value: ""
    };
    return _this;
  }

  _createClass(UiLabel, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.setVal();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      if (!this.state.value) this.setVal();
    }
  }, {
    key: 'openLink',
    value: function openLink(item) {
      window.open(item.hyperLink + "/" + encodeURIComponent(this.state.value || this.props.getVal(item.jsonPath, item.isDate)), 'mywin', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0');
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        null,
        this.renderLabel(this.props.item)
      );
    }
  }]);

  return UiLabel;
}(_react.Component);

exports.default = UiLabel;
