// egov-ui-react-framework/components/UiMultiSelectField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _SelectField = require('material-ui/SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _MenuItem = require('material-ui/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _egovCommonUtility = require('egov-common-utility');

var _jsonpath = require('jsonpath');

var _jsonpath2 = _interopRequireDefault(_jsonpath);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiMultiSelectField = function (_Component) {
	_inherits(UiMultiSelectField, _Component);

	function UiMultiSelectField(props) {
		_classCallCheck(this, UiMultiSelectField);

		var _this = _possibleConstructorReturn(this, (UiMultiSelectField.__proto__ || Object.getPrototypeOf(UiMultiSelectField)).call(this, props));

		_this.renderMultiSelect = function (item) {
			var dropDownData = _this.props.dropDownData;

			switch (_this.props.ui) {
				case 'google':
					return _react2.default.createElement(
						'div',
						{ style: { "display": "flex", "flex-direction": "column-reverse" } },
						_react2.default.createElement(
							_SelectField2.default,
							{
								className: 'custom-form-control-for-select',
								id: item.jsonPath.split(".").join("-"),
								floatingLabelStyle: { "color": item.isDisabled ? "#A9A9A9" : "#696969", "fontSize": "20px" },
								inputStyle: { "color": "#5F5C57" },
								floatingLabelFixed: true,
								dropDownMenuProps: { animated: false, targetOrigin: { horizontal: 'left', vertical: 'bottom' } },
								style: { "display": item.hide ? 'none' : 'block' },
								errorStyle: { "float": "left" },
								fullWidth: true,
								hintText: 'Please Select',
								multiple: true,
								floatingLabelText: _react2.default.createElement(
									'span',
									null,
									item.label,
									' ',
									_react2.default.createElement(
										'span',
										{ style: { "color": "#FF0000" } },
										item.isRequired ? " *" : ""
									)
								),
								value: _this.props.getVal(item.jsonPath),
								disabled: item.isDisabled,
								onChange: function onChange(ev, key, val) {
									_this.props.handler({ target: { value: val } }, item.jsonPath, item.isRequired ? true : false, '', item.requiredErrMsg, item.patternErrMsg, item.expression, item.expressionMsg);
								},
								errorText: _this.props.fieldErrors[item.jsonPath],
								maxHeight: 200 },
							dropDownData.hasOwnProperty(item.jsonPath) && dropDownData[item.jsonPath].map(function (dd, index) {
								return _react2.default.createElement(_MenuItem2.default, { value: dd.key, key: index, primaryText: dd.value });
							})
						)
					);
			}
		};

		_this.state = {
			dropDownData: []
		};
		return _this;
	}

	_createClass(UiMultiSelectField, [{
		key: 'componentWillReceiveProps',
		value: function componentWillReceiveProps(nextProps, nextState) {
			if (!_lodash2.default.isEqual(nextProps, this.props)) {
				this.initDat(nextProps);
			}
		}
	}, {
		key: 'initDat',
		value: function initDat(props) {
			var item = props.item,
			    setDropDownData = props.setDropDownData,
			    useTimestamp = props.useTimestamp;

			if (item.hasOwnProperty("url") && item.url.search("\\|") > -1) {
				var splitArray = item.url.split("?");
				var context = "";
				var id = {};
				for (var j = 0; j < splitArray[0].split("/").length; j++) {
					context += splitArray[0].split("/")[j] + "/";
				}

				var queryStringObject = splitArray[1].split("|")[0].split("&");
				for (var i = 0; i < queryStringObject.length; i++) {
					if (i) {
						id[queryStringObject[i].split("=")[0]] = queryStringObject[i].split("=")[1];
					}
				}

				_egovCommonUtility.Api.commonApiPost(context, id, {}, "", useTimestamp || false, false, "", "", item.isStateLevel).then(function (response) {
					if (response) {
						var keys = _jsonpath2.default.query(response, splitArray[1].split("|")[1]);
						var values = _jsonpath2.default.query(response, splitArray[1].split("|")[2]);
						var dropDownData = [];
						for (var k = 0; k < keys.length; k++) {
							var obj = {};
							obj["key"] = item.convertToString ? keys[k].toString() : item.convertToNumber ? Number(keys[k]) : keys[k];
							obj["value"] = values[k];
							dropDownData.push(obj);
						}

						dropDownData.sort(function (s1, s2) {
							return s1.value < s2.value ? -1 : s1.value > s2.value ? 1 : 0;
						});

						dropDownData.unshift({ key: null, value: "-- Please Select --" });
						setDropDownData(item.jsonPath, dropDownData);
					}
				}, function (err) {
					console.log(err);
				});
			}
		}
	}, {
		key: 'componentDidMount',
		value: function componentDidMount() {
			this.initDat(this.props);
		}
	}, {
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				null,
				this.renderMultiSelect(this.props.item)
			);
		}
	}]);

	return UiMultiSelectField;
}(_react.Component);

var mapStateToProps = function mapStateToProps(state) {
	return { dropDownData: state.framework.dropDownData, formData: state.frameworkForm.form };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
	return {
		setDropDownData: function setDropDownData(fieldName, dropDownData) {
			dispatch({ type: "SET_DROPDWON_DATA", fieldName: fieldName, dropDownData: dropDownData });
		}
	};
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(UiMultiSelectField);
