// egov-ui-react-framework/components/UiPayment.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrap = require('react-bootstrap');

var _Card = require('material-ui/Card');

var _SelectField = require('material-ui/SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _MenuItem = require('material-ui/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _TextField = require('material-ui/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _DatePicker = require('material-ui/DatePicker');

var _DatePicker2 = _interopRequireDefault(_DatePicker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var style = {
    margin: 12
};

var CashOrMops = function CashOrMops() {
    return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(_TextField2.default, {
                className: 'custom-form-control-for-textfield',
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelFixed: true,
                floatingLabelText: 'Amount:'
            }),
            _react2.default.createElement('br', null),
            _react2.default.createElement(_TextField2.default, {
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelText: 'Paid By:'
            })
        )
    );
};

var ChequeOrDD = function ChequeOrDD() {
    return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(_TextField2.default, {
                    className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'DD/Cheque Number:'
                })
            ),
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(_DatePicker2.default, { className: 'custom-form-control-for-textfield', floatingLabelStyle: { "color": "#696969", "fontSize": "20px" }, floatingLabelText: 'DD/Cheque Date', container: 'inline' })
            )
        ),
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(_TextField2.default, {
                    className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'Bank Name:'
                })
            ),
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(_TextField2.default, {
                    className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'Branch Name:'
                })
            )
        ),
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelFixed: true,
                floatingLabelText: 'Amount:'
            }),
            _react2.default.createElement('br', null),
            _react2.default.createElement(_TextField2.default, {
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelText: 'Paid By:'
            })
        )
    );
};

var CreditOrDebitCard = function CreditOrDebitCard() {
    return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 4 },
                _react2.default.createElement(_TextField2.default, {
                    className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'Last four digits of card:'
                })
            ),
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 4 },
                _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'Transacttion Number:'
                })
            ),
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 4 },
                _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'Re-enter Transacttion Number:'
                })
            )
        ),
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelFixed: true,
                floatingLabelText: 'Amount:'
            }),
            _react2.default.createElement('br', null),
            _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelFixed: true,
                floatingLabelText: 'Paid By:'
            })
        )
    );
};

var DirectBank = function DirectBank(props) {
    return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                    floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                    floatingLabelFixed: true,
                    floatingLabelText: 'Challan/Reference No'
                })
            ),
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(_DatePicker2.default, { className: 'custom-form-control-for-textfield', floatingLabelStyle: { "color": "#696969", "fontSize": "20px" }, floatingLabelText: 'Challan/Transaction Date', container: 'inline' })
            )
        ),
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(
                    _SelectField2.default,
                    { className: 'custom-form-control-for-select',
                        floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                        floatingLabelFixed: true,
                        floatingLabelText: 'Bank Name',
                        value: props.value.valueBankName,
                        onChange: props.handleChange
                    },
                    _react2.default.createElement(_MenuItem2.default, { value: "null", primaryText: '--Select--' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "bank1", primaryText: 'bank1' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "bank2", primaryText: 'bank2' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "bank3", primaryText: 'bank3' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "bank4", primaryText: 'bank4' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "bank5", primaryText: 'bank5' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "bank6", primaryText: 'bank6' })
                )
            ),
            _react2.default.createElement(
                _reactBootstrap.Col,
                { xs: 12, md: 6 },
                _react2.default.createElement(
                    _SelectField2.default,
                    { className: 'custom-form-control-for-select',
                        floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                        floatingLabelFixed: true,
                        floatingLabelText: 'Account Number',
                        value: props.value.valueActNum,
                        onChange: function onChange(e, index, value) {
                            var e = {
                                target: {
                                    value: value
                                }

                            };
                            props.handleChange(e, "valueBankName");
                        }
                    },
                    _react2.default.createElement(_MenuItem2.default, { value: "null", primaryText: '--Select--' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "acc1", primaryText: 'acc1' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "acc2", primaryText: 'acc2' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "acc3", primaryText: 'acc3' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "acc4", primaryText: 'acc4' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "acc5", primaryText: 'acc5' }),
                    _react2.default.createElement(_MenuItem2.default, { value: "acc6", primaryText: 'acc6' })
                )
            )
        ),
        _react2.default.createElement(
            _reactBootstrap.Col,
            { xs: 12, md: 12 },
            _react2.default.createElement(_TextField2.default, { className: 'custom-form-control-for-textfield',
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelFixed: true,
                floatingLabelText: 'Amount:'
            }),
            _react2.default.createElement('br', null),
            _react2.default.createElement(_TextField2.default, {
                floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
                floatingLabelText: 'Paid By:'
            })
        )
    );
};

var CollectionsNew = function (_Component) {
    _inherits(CollectionsNew, _Component);

    function CollectionsNew() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, CollectionsNew);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = CollectionsNew.__proto__ || Object.getPrototypeOf(CollectionsNew)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
            dropDown: {
                modeOfPayment: null,
                valueBankName: null,
                valueActNum: null
            }
        }, _this.handleChange = function (event, property) {
            _this.setState({ dropdown: _extends({}, _this.state.dropdown, _defineProperty({}, property, event.target.value))
            });
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(CollectionsNew, [{
        key: 'render',
        value: function render() {
            var _this2 = this;

            console.log(this.state.dropDown);
            var modeOfPayment = this.state.dropDown.modeOfPayment;

            return _react2.default.createElement(
                'div',
                null,
                _react2.default.createElement(
                    'card',
                    null,
                    _react2.default.createElement(
                        _reactBootstrap.Col,
                        { xs: 12, md: 6 },
                        _react2.default.createElement(
                            'h4',
                            null,
                            'Total Amount To Be Received:'
                        )
                    ),
                    _react2.default.createElement(
                        _reactBootstrap.Col,
                        { xs: 12, md: 6 },
                        _react2.default.createElement(
                            'h4',
                            null,
                            'Total Amount Received:'
                        )
                    ),
                    _react2.default.createElement(
                        _reactBootstrap.Col,
                        { xs: 12, md: 12 },
                        _react2.default.createElement(
                            _SelectField2.default,
                            {
                                floatingLabelFixed: true,
                                floatingLabelText: 'How would you like to pay?',
                                value: modeOfPayment ? modeOfPayment : null,
                                onChange: function onChange(e, index, value) {
                                    var event = {
                                        target: {
                                            value: value
                                        }

                                    };
                                    _this2.handleChange(event, "modeOfPayment");
                                }
                            },
                            '>',
                            _react2.default.createElement(_MenuItem2.default, { value: null, primaryText: '' }),
                            _react2.default.createElement(_MenuItem2.default, { value: 1, primaryText: 'Cash' }),
                            _react2.default.createElement(_MenuItem2.default, { value: 2, primaryText: 'Cheque' }),
                            _react2.default.createElement(_MenuItem2.default, { value: 3, primaryText: 'DD' }),
                            _react2.default.createElement(_MenuItem2.default, { value: 4, primaryText: 'Credit/Debit Card' }),
                            _react2.default.createElement(_MenuItem2.default, { value: 5, primaryText: 'Direct Bank' }),
                            _react2.default.createElement(_MenuItem2.default, { value: 6, primaryText: 'SBI MOPS Bank Callan' })
                        )
                    ),
                    this.state.dropDown.modeOfPayment == 1 && _react2.default.createElement(CashOrMops, null),
                    this.state.dropDown.modeOfPayment == 2 && _react2.default.createElement(ChequeOrDD, null),
                    this.state.dropDown.modeOfPayment == 3 && _react2.default.createElement(ChequeOrDD, null),
                    this.state.dropDown.modeOfPayment == 4 && _react2.default.createElement(CreditOrDebitCard, null),
                    this.state.dropDown.modeOfPayment == 5 && _react2.default.createElement(DirectBank, { value: this.state.dropdown, handleChange: this.state.handleChange }),
                    this.state.dropDown.modeOfPayment == 6 && _react2.default.createElement(CashOrMops, null),
                    _react2.default.createElement(
                        _reactBootstrap.Col,
                        { xs: 12, md: 12 },
                        _react2.default.createElement(
                            _reactBootstrap.Col,
                            { xs: 12, md: 2 },
                            _react2.default.createElement(_RaisedButton2.default, { label: 'Pay', primary: true, style: style })
                        ),
                        _react2.default.createElement(
                            _reactBootstrap.Col,
                            { xs: 12, md: 2 },
                            _react2.default.createElement(_RaisedButton2.default, { label: 'Option2', secondary: true, style: style })
                        ),
                        _react2.default.createElement(
                            _reactBootstrap.Col,
                            { xs: 12, md: 2 },
                            _react2.default.createElement(_RaisedButton2.default, { label: 'Option3', secondary: true, style: style })
                        )
                    ),
                    _react2.default.createElement(
                        'h4',
                        null,
                        'Label3'
                    )
                )
            );
        }
    }]);

    return CollectionsNew;
}(_react.Component);

exports.default = CollectionsNew;
