// egov-ui-react-framework/components/UiWorkflow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TextField = require('material-ui/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _SelectField = require('material-ui/SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _MenuItem = require('material-ui/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _reactBootstrap = require('react-bootstrap');

var _egovCommonUtility = require('egov-common-utility');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UiWorkflow = function (_Component) {
	_inherits(UiWorkflow, _Component);

	function UiWorkflow(props) {
		_classCallCheck(this, UiWorkflow);

		var _this = _possibleConstructorReturn(this, (UiWorkflow.__proto__ || Object.getPrototypeOf(UiWorkflow)).call(this, props));

		_initialiseProps.call(_this);

		_this.state = {
			history: [],
			departments: [],
			buttons: [],
			employees: [],
			hide: false,
			status: ''
		};
		return _this;
	}

	_createClass(UiWorkflow, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			this.initData(this.props);
		}
	}, {
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				null,
				this.renderUiWorkflow()
			);
		}
	}]);

	return UiWorkflow;
}(_react.Component);

var _initialiseProps = function _initialiseProps() {
	var _this2 = this;

	this.initData = function (props) {
		var item = props.item,
		    workflowId = props.workflowId;


		_egovCommonUtility.Api.commonApiPost("egov-common-masters/departments/_search", {}, {}, null, false).then(function (res) {
			this.setState({
				departments: res && res.Department && res.Department.constructor == Array ? res.Department : []
			});
		}, function (err) {});

		if (workflowId) {
			//Fetch workflow history
			_egovCommonUtility.Api.commonApiPost("egov-common-workflows/history", {
				workflowId: workflowId
			}, {}, null, true).then(function (res) {
				this.setState({
					history: res && res.tasks && res.tasks.constructor == Array ? res.tasks : []
				});
			}, function (err) {});

			//Fetch buttons
			_egovCommonUtility.Api.commonApiPost("egov-common-workflows/process/_search", {
				id: workflowId
			}, {}, null, false).then(function (res) {
				if (res && res.processInstance && res.processInstance.attributes && res.processInstance.attributes.validActions && res.processInstance.attributes.validActions.values && res.processInstance.attributes.validActions.values.length) {
					var flg = 0;
					for (var j = 0; j < res.processInstance.attributes.validActions.values.length; j++) {
						if (res.processInstance.attributes.validActions.values[j].key.toLowerCase() == "forward" || res.processInstance.attributes.validActions.values[j].key.toLowerCase() == "submit") {
							flg = 1;
						}
					}

					this.setState({
						buttons: res.processInstance.attributes.validActions.values,
						hide: flg == 1 ? false : true
					});
				}

				if (res && res.processInstance) {
					_egovCommonUtility.Api.commonApiPost("/egov-common-workflows/designations/_search", {
						businessKey: item.businessKey,
						approvalDepartmentName: "",
						departmentRule: "",
						currentStatus: res.processInstance.status,
						additionalRule: "",
						pendingAction: "",
						designation: "",
						amountRule: ""
					}, {}, null, false).then(function (res3) {
						if (res3 && res3.length) {
							var count = res3.length;

							var _loop = function _loop(i) {
								_egovCommonUtility.Api.commonApiPost("/hr-masters/designations/_search", {
									name: res3[i].name
								}, {}, null, false).then(function (res2) {
									res3[i].id = res2.Designation && res2.Designation[0] ? res2.Designation[0].id : "-";
									count--;
									if (count == 0) {
										this.setState({
											designations: res3,
											status: res.processInstance.status
										});
									}
								}, function (err) {});
							};

							for (var i = 0; i < res3.length; i++) {
								_loop(i);
							}
						}
					}, function (err) {});
				}
			}, function (err) {});
		} else {
			_egovCommonUtility.Api.commonApiPost("/egov-common-workflows/designations/_search", {
				businessKey: item.businessKey,
				approvalDepartmentName: "",
				departmentRule: "",
				currentStatus: "",
				additionalRule: "",
				pendingAction: "",
				designation: "",
				amountRule: ""
			}, {}, null, false).then(function (res3) {
				if (res3 && res3.length) {
					var count = res3.length;

					var _loop2 = function _loop2(i) {
						_egovCommonUtility.Api.commonApiPost("/hr-masters/designations/_search", {
							name: res3[i].name
						}, {}, null, false).then(function (res2) {
							res3[i].id = res2.Designation && res2.Designation[0] ? res2.Designation[0].id : "-";
							count--;
							if (count == 0) {
								this.setState({
									designations: res3
								});
							}
						}, function (err) {});
					};

					for (var i = 0; i < res3.length; i++) {
						_loop2(i);
					}
				}
			}, function (err) {});
		}
	};

	this.getEmployee = function (item) {
		if (_this2.props.getVal(item.jsonPath.departmentPath) && _this2.props.getVal(item.jsonPath.designationPath)) {
			_egovCommonUtility.Api.commonApiPost("hr-employee/employees/_search", {
				departmentId: _this2.props.getVal(item.jsonPath.departmentPath),
				designationId: _this2.props.getVal(item.jsonPath.designationPath)
			}, {}, null, false).then(function (res) {
				this.setState({
					employees: res.Employee
				});
			}, function (err) {});
		}
	};

	this.getPosition = function (id) {
		for (var i = 0; i < _this2.state.employees.length; i++) {
			if (_this2.state.employees[i].id == id) {
				if (_this2.state.employees[i].assignments && _this2.state.employees[i].assignments[0]) {
					return _this2.state.employees[i].assignments[0].position;
				}
				break;
			}
		}

		return "";
	};

	this.renderWFHistory = function (item) {
		var history = _this2.state.history;

		if (history && history.length) {
			return _react2.default.createElement(
				_reactBootstrap.Table,
				{ bordered: true, responsive: true, className: 'table-striped' },
				_react2.default.createElement(
					'thead',
					null,
					_react2.default.createElement(
						'tr',
						null,
						_react2.default.createElement(
							'th',
							null,
							(0, _egovCommonUtility.translate)("employee.ServiceHistory.fields.date")
						),
						_react2.default.createElement(
							'th',
							null,
							(0, _egovCommonUtility.translate)("wc.create.workflow.UpdatedBy")
						),
						_react2.default.createElement(
							'th',
							null,
							(0, _egovCommonUtility.translate)("collection.create.status")
						),
						_react2.default.createElement(
							'th',
							null,
							(0, _egovCommonUtility.translate)("wc.create.workflow.currentOwner")
						),
						_react2.default.createElement(
							'th',
							null,
							(0, _egovCommonUtility.translate)("reports.common.comments")
						)
					)
				),
				_react2.default.createElement(
					'tbody',
					null,
					history.map(function (v, i) {
						return _react2.default.createElement(
							'tr',
							{ key: i },
							_react2.default.createElement(
								'td',
								null,
								v.createdDate || "-"
							),
							_react2.default.createElement(
								'td',
								null,
								v.senderName || "-"
							),
							_react2.default.createElement(
								'td',
								null,
								v.status || "-"
							),
							_react2.default.createElement(
								'td',
								null,
								v.owner && v.owner.name ? v.owner.name : "-"
							),
							_react2.default.createElement(
								'td',
								null,
								v.comments || "-"
							)
						);
					})
				)
			);
		}
	};

	this.renderWorkflowCard = function (item) {
		return _react2.default.createElement(
			'div',
			null,
			!_this2.state.hide && _react2.default.createElement(
				_reactBootstrap.Row,
				null,
				_react2.default.createElement(
					_reactBootstrap.Col,
					{ xs: 12, md: 3 },
					_react2.default.createElement(
						_SelectField2.default,
						{
							className: 'custom-form-control-for-select',
							dropDownMenuProps: { animated: true, targetOrigin: { horizontal: 'left', vertical: 'bottom' } },
							floatingLabelFixed: true,
							floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
							floatingLabelText: _react2.default.createElement(
								'span',
								null,
								(0, _egovCommonUtility.translate)("employee.Assignment.fields.department"),
								_react2.default.createElement(
									'span',
									{ style: { "color": "#FF0000" } },
									' *'
								)
							),
							value: _this2.props.getVal(item.jsonPath.departmentPath),
							onChange: function onChange(event, key, value) {
								_this2.getEmployee(item);
								_this2.props.handler({ target: { value: value } }, item.jsonPath.departmentPath, true, item.requiredErrMsg, item.patternErrMsg);
							} },
						_this2.state.departments && _this2.state.departments.map(function (v, i) {
							return _react2.default.createElement(_MenuItem2.default, { value: v.id, key: i, primaryText: v.name });
						})
					)
				),
				_react2.default.createElement(
					_reactBootstrap.Col,
					{ xs: 12, md: 3 },
					_react2.default.createElement(
						_SelectField2.default,
						{
							className: 'custom-form-control-for-select',
							floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
							floatingLabelFixed: true,
							dropDownMenuProps: { animated: true, targetOrigin: { horizontal: 'left', vertical: 'bottom' } },
							floatingLabelText: _react2.default.createElement(
								'span',
								null,
								(0, _egovCommonUtility.translate)("employee.Assignment.fields.designation"),
								_react2.default.createElement(
									'span',
									{ style: { "color": "#FF0000" } },
									' *'
								)
							),
							value: _this2.props.getVal(item.jsonPath.designationPath),
							onChange: function onChange(event, key, value) {
								_this2.getEmployee(item);
								_this2.props.handler({ target: { value: value } }, item.jsonPath.designationPath, true, item.requiredErrMsg, item.patternErrMsg);
							} },
						_this2.state.designations && _this2.state.designations.map(function (v, i) {
							return _react2.default.createElement(_MenuItem2.default, { value: v.id, key: i, primaryText: v.name });
						})
					)
				),
				_react2.default.createElement(
					_reactBootstrap.Col,
					{ xs: 12, md: 3 },
					_react2.default.createElement(
						_SelectField2.default,
						{
							className: 'custom-form-control-for-select',
							floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
							floatingLabelFixed: true,
							dropDownMenuProps: { animated: true, targetOrigin: { horizontal: 'left', vertical: 'bottom' } },
							floatingLabelText: _react2.default.createElement(
								'span',
								null,
								(0, _egovCommonUtility.translate)("wc.create.groups.approvalDetails.fields.approver"),
								_react2.default.createElement(
									'span',
									{ style: { "color": "#FF0000" } },
									' *'
								)
							),
							value: _this2.props.getVal(item.jsonPath.assigneePath),
							onChange: function onChange(event, key, value) {
								_this2.props.handler({ target: { value: _this2.getPosition(value) } }, item.jsonPath.assigneePath, true, item.requiredErrMsg, item.patternErrMsg);
							} },
						_this2.state.employees && _this2.state.employees.map(function (v, i) {
							return _react2.default.createElement(_MenuItem2.default, { value: v.id, key: i, primaryText: v.name });
						})
					)
				)
			),
			_react2.default.createElement(
				_reactBootstrap.Row,
				null,
				_react2.default.createElement(
					_reactBootstrap.Col,
					{ xs: 12, md: 12 },
					_react2.default.createElement(_TextField2.default, {
						className: 'custom-form-control-for-textfield',
						floatingLabelStyle: { "color": "#696969", "fontSize": "20px" },
						floatingLabelFixed: true,
						type: 'text',
						multiple: true,
						fullWidth: true,
						rows: 3,
						floatingLabelText: (0, _egovCommonUtility.translate)("wc.create.groups.approvalDetails.fields.comments"),
						value: _this2.props.getVal(item.jsonPath.commentsPath),
						onChange: function onChange(e) {
							_this2.props.handler(e, item.jsonPath.commentsPath, true, item.requiredErrMsg, item.patternErrMsg);
						} })
				)
			)
		);
	};

	this.renderButtons = function (item) {
		if (_this2.state.buttons && _this2.state.buttons.length) {
			return _react2.default.createElement(
				'div',
				{ style: { "textAlign": "center" } },
				_this2.state.buttons.map(function (v, i) {
					var _this3 = this;

					return _react2.default.createElement(
						'span',
						null,
						_react2.default.createElement(_RaisedButton2.default, {
							onClick: function onClick(e) {
								_this3.props.initiateWF(v, item, _this3.state.hide, _this3.state.status);
							},
							label: v.name,
							primary: true }),
						'\xA0\xA0'
					);
				})
			);
		}
	};

	this.renderUiWorkflow = function (item) {
		switch (_this2.props.ui) {
			case 'google':
				return _react2.default.createElement(
					'div',
					null,
					_this2.renderWFHistory(item),
					_react2.default.createElement('br', null),
					_this2.renderWorkflowCard(item),
					_react2.default.createElement('br', null),
					_this2.renderButtons(item)
				);
		}
	};
};

exports.default = UiWorkflow;
