// egov-ui-react-framework/templates/templateParser/templateParser.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _index = require('../templates/index');

var Templates = _interopRequireWildcard(_index);

var _UiBackButton = require('../../components/UiBackButton');

var _UiBackButton2 = _interopRequireDefault(_UiBackButton);

var _reactRedux = require('react-redux');

var _reactRouter = require('react-router');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TemplateParser = function (_Component) {
	_inherits(TemplateParser, _Component);

	function TemplateParser(props) {
		_classCallCheck(this, TemplateParser);

		var _this = _possibleConstructorReturn(this, (TemplateParser.__proto__ || Object.getPrototypeOf(TemplateParser)).call(this, props));

		_this.printTemplate = function () {
			var mywindow = window.open('', 'PRINT', 'height=400,width=600');
			var cdn = '\n\t      <!-- Latest compiled and minified CSS -->\n\t      <link rel="stylesheet" media="all" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">\n\n\t      <!-- Optional theme -->\n\t      <link rel="stylesheet" media="all" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">  ';
			mywindow.document.write('<html><head><title> </title>');
			mywindow.document.write(cdn);
			mywindow.document.write('</head><body>');
			mywindow.document.write(document.getElementById('printTemplate').innerHTML);
			mywindow.document.write('</body></html>');

			mywindow.document.close(); // necessary for IE >= 10
			mywindow.focus(); // necessary for IE >= 10*/

			setTimeout(function () {
				mywindow.print();
				mywindow.close();
			}, 1000);
		};

		_this.back = function (e) {
			var setRoute = _this.props.setRoute;

			setRoute("/" + localStorage.getItem("returnUrl"));
		};

		return _this;
	}

	_createClass(TemplateParser, [{
		key: 'render',
		value: function render() {
			var Template = Templates[this.props.match.params.templatePath];
			var back = this.back;

			return _react2.default.createElement(
				'div',
				null,
				_react2.default.createElement(
					'div',
					{ id: 'printTemplate' },
					localStorage.reportData ? JSON.parse(localStorage.reportData).map(function (v, i) {
						return _react2.default.createElement(
							'div',
							null,
							_react2.default.createElement(
								'div',
								{ style: { "textAlign": "right",
										"paddingRight": "15px" } },
								_react2.default.createElement('br', null),
								_react2.default.createElement(_RaisedButton2.default, { type: 'button', onClick: function onClick(e) {
										back(e);
									}, primary: true, label: "Back" }),
								_react2.default.createElement('br', null),
								_react2.default.createElement('br', null)
							),
							_react2.default.createElement(Template, { data: v }),
							_react2.default.createElement('br', null),
							_react2.default.createElement('div', { style: { "page-break-after": "always" } })
						);
					}) : ""
				),
				_react2.default.createElement(
					'div',
					{ style: { "textAlign": "center" } },
					_react2.default.createElement(_RaisedButton2.default, { label: "Print", primary: true, onClick: this.printTemplate })
				)
			);
		}
	}]);

	return TemplateParser;
}(_react.Component);

var mapStateToProps = function mapStateToProps(state) {
	return {};
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
	return {
		setRoute: function setRoute(route) {
			return dispatch({ type: "SET_ROUTE", route: route });
		}
	};
};

exports.default = (0, _reactRouter.withRouter)((0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(TemplateParser));
